--- subroot-HRM/subroot-HRM.b256a8d.edit-HRM.mc	2026-03-23 13:07:59.134363781 +0100
+++ subroot-HRM/subroot-HRM.b256a8d.renamed-HRM.mc	2026-03-23 12:45:21.769480764 +0100
@@ -2245,22 +2245,6 @@
                             exit 1)
                          1)
                   in
-                  -- Export leaf messages (interactions)
-                  let ll = (match tree1 with Leaf carried2
-                          then
-                            carried2.label
-                          else match tree1 with Node carried3
-                          then
-                            carried3.label
-                          else
-                            let #var"22" =
-                              print
-                                "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/host-rep-lib/belief-propagation.tppl 9:32-9:42>:\nField \'label\' not found\n[0m    let leafInts = interactions[[31mtree.label[0m[0m];\n"
-                            in
-                            exit 1) in
-                  let leafMsg = observationMessage leafInts in
-                  let st40 = p_export st40 (hrmStoreNodeMsg ll) (p_pure leafMsg) in
-                  -- Done 
                   (st40, MsgLeaf
                     { age = 0.,
                       label =
@@ -2428,18 +2412,8 @@
                 with
                   (st50, rightInMsg)
                 in
-                match mtxElemMul st50 leftInMsg rightInMsg with (st51, x50) 
+                match mtxElemMul st50 leftInMsg rightInMsg with (st51, x50)
                 in
-                -- Export transition kernels
-                let getLabel = lam n.
-                  match n with MsgLeaf l then l.label
-                  else match n with MsgNode n in n.label
-                in
-                let leftLabel = getLabel left in
-                let rightLabel = getLabel right in
-                let st51 = p_export st51 (hrmStoreTransKernel rightLabel) rightKernel in 
-                let st51 = p_export st51 (hrmStoreTransKernel leftLabel) leftKernel in 
-                -- Done
                 (st51, MsgNode
                   { age =
                       match tree1 with Leaf carried18
@@ -6172,7 +6146,7 @@
             with
               (st613, x953)
             in
-            p_assume st613 (hrmStoreNode nodeLabel51 x947) x953
+            p_assume st613 hrmStoreAssume x953
   in
   recursive
     let rec15 =
@@ -6337,7 +6311,7 @@
                                in
                                match p_apply st293 x433 x435 with (st294, x436)
                                in
-                               match p_assume st294 (hrmStoreBranchState nodeLabel host) x436 with (st295, nextState11)
+                               match p_assume st294 hrmStoreAssume x436 with (st295, nextState11)
                                in
                                match
                                  p_traverseSeq
@@ -6384,7 +6358,7 @@
                                in
                                match p_apply st299 x442 x444 with (st300, x445)
                                in
-                               match p_assume st300 (hrmStoreBranchTime nodeLabel host) x445 with (st301, x446)
+                               match p_assume st300 hrmStoreAssume x445 with (st301, x446)
                                in
                                match
                                  p_map
@@ -6574,7 +6548,7 @@
                                in
                                match p_apply st321 x479 x481 with (st322, x482)
                                in
-                               match p_assume st322 (hrmStoreBranchState nodeLabel1 host1) x482 with (st323, nextState12)
+                               match p_assume st322 hrmStoreAssume x482 with (st323, nextState12)
                                in
                                match
                                  p_traverseSeq
@@ -6621,7 +6595,7 @@
                                in
                                match p_apply st327 x488 x490 with (st328, x491)
                                in
-                               match p_assume st328 (hrmStoreBranchTime nodeLabel1 host1) x491 with (st329, x492)
+                               match p_assume st328 hrmStoreAssume x491 with (st329, x492)
                                in
                                match
                                  p_map
@@ -6807,7 +6781,7 @@
                     in
                     match p_apply st599 x934 x936 with (st600, x937)
                     in
-                    match p_assume st600 (hrmStoreBranchTime nodeLabel13 x928) x937 with (st601, x938)
+                    match p_assume st600 hrmStoreAssume x937 with (st601, x938)
                     in
                     match
                       p_map
@@ -6853,7 +6827,7 @@
                     let st606 =
                       p_weight
                         st605
-                        (hrmStoreBridgeSuppWeight nodeLabel13 x928)
+                        hrmStoreWeight
                         (lam x945.
                            externalLog
                              (addf
@@ -6942,7 +6916,7 @@
                     in
                     match p_apply st587 x915 x917 with (st588, x918)
                     in
-                    match p_assume st588 (hrmStoreBranchTime nodeLabel12 x908) x918 with (st589, x919)
+                    match p_assume st588 hrmStoreAssume x918 with (st589, x919)
                     in
                     match
                       p_map
@@ -6988,7 +6962,7 @@
                     let st594 =
                       p_weight
                         st593
-                        (hrmStoreBridgeSuppWeight nodeLabel12 x908)
+                        hrmStoreWeight
                         (lam x927.
                            externalLog
                              (addf
@@ -8659,22 +8633,10 @@
                         in
                         match p_apply st431 x679 x681 with (st432, x682)
                         in
-                        let nl = (match tree2 with MsgLeaf carried24
-                                             then
-                                               carried24.label
-                                             else match tree2 with MsgNode carried25
-                                             then
-                                               carried25.label
-                                             else
-                                               let #var"263" =
-                                                 print
-                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/flat-root-prior-HRM.tppl 129:90-129:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
-                                               in
-                                               exit 1) in
                         let st433 =
                           p_weight
                             st432
-                            (hrmStoreLikrWeight nl)
+                            hrmStoreWeight
                             (/-temp-/lam x685.
                                x685)
                             x682
@@ -8771,22 +8733,10 @@
                       in
                       match bool2real st437 x686 with (st438, x687)
                       in
-                      let nl = (match tree2 with MsgLeaf carried50
-                                         then
-                                           carried50.label
-                                         else match tree2 with MsgNode carried51
-                                         then
-                                           carried51.label
-                                         else
-                                           let #var"293" =
-                                             print
-                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/flat-root-prior-HRM.tppl 141:43-141:53>:\nField \'label\' not found\n[0m    weight bool2real(any(is2, rep)) + Real([31mtree.label[0m[0m) * 0.;\n"
-                                           in
-                                           exit 1) in
                       let st439 =
                         p_weight
                           st438
-                          (hrmStoreNodeSuppWeight nl)
+                          hrmStoreWeight
                           (lam x708.
                              externalLog
                                (addf
@@ -8920,22 +8870,10 @@
                       in
                       match p_apply st445 x694 x697 with (st446, x698)
                       in
-                      let nl = (match tree2 with MsgLeaf carried40
-                                         then
-                                           carried40.label
-                                         else match tree2 with MsgNode carried41
-                                         then
-                                           carried41.label
-                                         else
-                                           let #var"276" =
-                                             print
-                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 211:113-211:123>:\nField \'label\' not found\n[0m    logWeight branchSample.logModelDensity - branchSample.logSamplingDensity - nodeLogSamplingDensity * exp(Real([31mtree.label[0m[0m) * 0.);\n"
-                                           in
-                                           exit 1) in
                       let st447 =
                         p_weight
                           st446
-                          (hrmStoreLikrWeight nl)
+                          hrmStoreWeight
                           (/-temp-/lam x707.
                              x707)
                           x698
@@ -8943,10 +8881,6 @@
                       let foo2 = {} in
                       match observationMessage1 st447 rep2 with (st448, newMsg)
                       in
-                      -- Export node messages
-                      let st448 = p_export st448 (hrmStoreNodeMsg nl) newMsg in
-                      let st448 = p_export st448 (hrmStorePreorderMsg nl) preorderMsg in
-                      -- Done
                       match
                         mtxMul
                           st448
@@ -9221,30 +9155,19 @@
   in
   let subrootLength = #var"X11" in
   let nHosts1 = length11 hostDistances in
-  -- Export the input tree
-  match input1 with {subrootLabel = #var"X13"}
-  in
-  let subrootLabel1 = #var"X13" in
-  recursive let buildTree = lam stree.
-    match stree with Leaf { label = label, age = _ } then 
-      HRMLeaf { label = label }
-    else match stree with Node { label = label, age = _, left = left, right = right } in 
-      HRMNode { label = label, left = buildTree left, right = buildTree right } in
-  match p_export st (hrmStoreTree (Some subrootLabel1) (buildTree symbiontTree) (match input1 with {interactions = ints} in ints)) (p_pure ()) with st in
-  -- Done
   match
     p_assume
-      st hrmStoreLambda (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
+      st hrmStoreAssume (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
   with
     (st454, lambda)
   in
   match
-    p_assume st454 hrmStoreMu (p_pure (mkExponential 10.))
+    p_assume st454 hrmStoreAssume (p_pure (mkExponential 10.))
   with
     (st455, mu)
   in
   match
-    p_assume st455 hrmStoreBeta (p_pure (mkExponential 1.))
+    p_assume st455 hrmStoreAssume (p_pure (mkExponential 1.))
   with
     (st456, beta)
   in
@@ -9468,6 +9391,9 @@
   with
     (st491, postorderTree)
   in
+  match input1 with {subrootLabel = #var"X13"}
+  in
+  let subrootLabel1 = #var"X13" in
   match mtxSclrMul st491 subrootLength qMatrix2 with (st492, x769)
   in
   match mtxExpRateSafe st492 x769 with (st493, subrootKernel)
@@ -9573,20 +9499,6 @@
   in
   match eitherEither st502 x782 with (st503, subrootSamplingProb)
   in
-  let st503 = p_export st503 (hrmStorePreorderMsg subrootLabel1) (p_pure b4) in
-  let rl = (match postorderTree with MsgLeaf carried56
-       then
-         carried56.label
-       else match postorderTree with MsgNode carried57
-       then
-         carried57.label
-       else
-         let #var"292" =
-           print
-             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 87:57-87:76>:\nField \'label\' not found\n[0m  let rootRep = suggestNodeRep(rootSamplingProb, nHosts, [31mpostorderTree.label[0m[0m);\n"
-         in
-         exit 1) in
-  let st503 = p_export st503 (hrmStoreTransKernel rl) subrootKernel in
   match
     suggestNodeRep st503 subrootSamplingProb nHosts1 subrootLabel1
   with
@@ -9599,7 +9511,7 @@
   let st507 =
     p_weight
       st506
-      (hrmStoreNodeSuppWeight subrootLabel1)
+      hrmStoreWeight
       (lam x811.
          externalLog (addf x811 (mulf (int2float subrootLabel1) 0.)))
       x784
@@ -9613,7 +9525,7 @@
   let st509 =
     p_weight
       st508
-      (hrmStoreLikrWeight subrootLabel1)
+      hrmStoreWeight
       (lam x810.
          addf
            (subf
@@ -9686,22 +9598,10 @@
   in
   match bool2real st514 x788 with (st515, x789)
   in
-  let rl = (match postorderTree with MsgLeaf carried56
-       then
-         carried56.label
-       else match postorderTree with MsgNode carried57
-       then
-         carried57.label
-       else
-         let #var"292" =
-           print
-             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 87:57-87:76>:\nField \'label\' not found\n[0m  let rootRep = suggestNodeRep(rootSamplingProb, nHosts, [31mpostorderTree.label[0m[0m);\n"
-         in
-         exit 1) in
   let st516 =
     p_weight
       st515
-      (hrmStoreNodeSuppWeight rl)
+      hrmStoreWeight
       (lam x809.
          externalLog
            (addf
@@ -9855,39 +9755,14 @@
   in
   match p_apply st522 x797 x800 with (st523, x801)
   in
-  let rl = (match postorderTree with MsgLeaf carried58
-                     then
-                       carried58.label
-                     else match postorderTree with MsgNode carried59
-                     then
-                       carried59.label
-                     else
-                       let #var"297" =
-                         print
-                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 110:40-110:59>:\nField \'label\' not found\n[0m    - rootLogSamplingDensity * exp(Real([31mpostorderTree.label[0m[0m) * 0.)\n"
-                       in
-                       exit 1) in
   let st524 =
     p_weight
-      st523 (hrmStoreLikrWeight rl) (/-temp-/lam x808.
+      st523 hrmStoreWeight (/-temp-/lam x808.
          x808) x801
   in
   let foo4 = {} in
   match observationMessage1 st524 rootRep with (st525, newMsg1)
   in
-  -- Export root messages
-  let st525 = p_export st525 (hrmStoreNodeMsg rl) newMsg1 in
-  let subrootMsg = (match subrootInfo with SubrootInfo1 x1091
-       then
-         x1091.msg
-       else
-         let #var"290" =
-           print
-             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 86:36-86:51>:\nField \'msg\' not found\n[0m  let rootSamplingProb = mtxElemMul([31msubrootInfo.msg[0m[0m, postorderTree.outMsg);\n"
-         in
-         p_pure (exit 1)) in
-  let st525 = p_export st525 (hrmStorePreorderMsg rl) subrootMsg in
-  -- Done
   let rootAge1 =
     match postorderTree with MsgLeaf carried66
     then
@@ -10009,48 +9884,44 @@
   with
     (st530, x807)
   in
-  -- Set up exports
-  let createExport = lam st. lam left. lam right.
-    let m = mapEmpty subi in
-    let makeDirtyArray = lam st. lam arr.
-      p_traverseSeq st (lam st. lam a. (st, a)) arr 
-    in
-    -- Traverse the tree to extract the node repertoires
-    recursive let extractSamples = lam st. lam m. lam hTree.
-      match hTree with HistoryLeaf _ then (st, m)
-      else match hTree with HistoryNode n in
-      match makeDirtyArray st n.repertoire with (st, wrappedRep) in
-      let m = mapInsert n.label wrappedRep m in
-      match extractSamples st m n.left with (st, m) in
-      extractSamples st m n.right
-    in
-    -- Save the root repertoire
-    match makeDirtyArray st rootRep with (st, dRootRep) in
-    let m = mapInsert rl dRootRep m in
-    -- Save the subroot repertoire
-    match makeDirtyArray st subrootRep with (st, dSubrootRep) in
-    let m = mapInsert subrootLabel1 dSubrootRep m in
-    -- Save the two subtrees
-    match extractSamples st m left with (st, m) in
-    match extractSamples st m right with (st, m) in
-    let dirtyKeyVal = lam st. lam kv. match kv with (k, v) in p_map st (lam v. (k, v)) v in 
-    match p_traverseSeq st dirtyKeyVal (mapToSeq m) with (st, wrappedReps) in
-    match p_map st (
-      lam mu. lam beta. lam lambda. lam reps.
-        -- printLn (join ["Export root rep: [", int2string rootRep1, ", ", int2string rootRep2, "]"]);
-        { mu = mu
-        , beta = beta
-        , lambda = lambda
-        , reps = mapFromSeq subi reps 
-        }
-    ) mu with (st, partres) in
-    match p_apply st partres beta with (st, partres) in
-    match p_apply st partres lambda with (st, partres) in
-    p_apply st partres wrappedReps
-  in
-  match createExport st530 x803 x805 with (st530, res) in
-  -- Done
   p_export
     st530
     hrmStoreExport
-    res
\ No newline at end of file
+    (p_pure
+       (dprint
+          (ReturnType1
+             { mu = mu,
+               beta = beta,
+               tree =
+                 HistoryNode
+                   { age =
+                       match symbiontTree with Leaf carried62
+                       then
+                         carried62.age
+                       else match symbiontTree with Node carried63
+                       then
+                         carried63.age
+                       else
+                         let #var"2106" =
+                           print
+                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 128:10-128:26>:\nField \'age\' not found\n[0m    age = [31msymbiontTree.age[0m[0m, label = symbiontTree.label,\n"
+                         in
+                         exit 1,
+                     label =
+                       match symbiontTree with Leaf carried64
+                       then
+                         carried64.label
+                       else match symbiontTree with Node carried65
+                       then
+                         carried65.label
+                       else
+                         let #var"2107" =
+                           print
+                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 128:36-128:54>:\nField \'label\' not found\n[0m    age = symbiontTree.age, label = [31msymbiontTree.label[0m[0m,\n"
+                         in
+                         exit 1,
+                     left = x803,
+                     right = x805,
+                     history = x807,
+                     repertoire = rootRep },
+               lambda = lambda })))
