--- subroot-HRM/subroot-HRM.d8222d5.edit-HRM.mc	2026-03-09 21:17:53.299319265 +0100
+++ subroot-HRM/subroot-HRM.d8222d5.renamed-HRM.mc	2026-03-09 20:24:16.339148049 +0100
@@ -2245,22 +2245,6 @@
                             exit 1)
                          1)
                   in
-                  -- Export leaf messages (interactions)
-                  let ll = (match tree1 with Leaf carried2
-                          then
-                            carried2.label
-                          else match tree1 with Node carried3
-                          then
-                            carried3.label
-                          else
-                            let #var"22" =
-                              print
-                                "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/host-rep-lib/belief-propagation.tppl 9:32-9:42>:\nField \'label\' not found\n[0m    let leafInts = interactions[[31mtree.label[0m[0m];\n"
-                            in
-                            exit 1) in
-                  let leafMsg = observationMessage leafInts in
-                  let st40 = p_export st40 (hrmStoreNodeMsg ll) (p_pure leafMsg) in
-                  -- Done 
                   (st40, MsgLeaf
                     { age = 0.,
                       label =
@@ -2428,18 +2412,7 @@
                 with
                   (st50, rightInMsg)
                 in
-                match mtxElemMul st50 leftInMsg rightInMsg with (st51, x50) 
-                in
-                -- Export transition kernels
-                let getLabel = lam n.
-                  match n with MsgLeaf l then l.label
-                  else match n with MsgNode n in n.label
-                in
-                let leftLabel = getLabel left in
-                let rightLabel = getLabel right in
-                let st51 = p_export st51 (hrmStoreTransKernel rightLabel) rightKernel in 
-                let st51 = p_export st51 (hrmStoreTransKernel leftLabel) leftKernel
-                -- Done
+                match mtxElemMul st50 leftInMsg rightInMsg with (st51, x50)
                 in
                 (st51, MsgNode
                   { age =
@@ -6070,7 +6043,7 @@
             with
               (st608, x946)
             in
-            p_assume st608 (hrmStoreNode nodeLabel51 x941) x946
+            p_assume st608 hrmStoreAssume x946
   in
   recursive
     let rec15 =
@@ -6235,7 +6208,7 @@
                                in
                                match p_apply st288 x427 x429 with (st289, x430)
                                in
-                               match p_assume st289 (hrmStoreBranchState nodeLabel host) x430 with (st290, nextState11)
+                               match p_assume st289 hrmStoreAssume x430 with (st290, nextState11)
                                in
                                match
                                  p_traverseSeq
@@ -6282,7 +6255,7 @@
                                in
                                match p_apply st294 x436 x438 with (st295, x439)
                                in
-                               match p_assume st295 (hrmStoreBranchTime nodeLabel host) x439 with (st296, x440)
+                               match p_assume st295 hrmStoreAssume x439 with (st296, x440)
                                in
                                match
                                  p_map
@@ -6472,7 +6445,7 @@
                                in
                                match p_apply st316 x473 x475 with (st317, x476)
                                in
-                               match p_assume st317 (hrmStoreBranchState nodeLabel1 host1) x476 with (st318, nextState12)
+                               match p_assume st317 hrmStoreAssume x476 with (st318, nextState12)
                                in
                                match
                                  p_traverseSeq
@@ -6519,7 +6492,7 @@
                                in
                                match p_apply st322 x482 x484 with (st323, x485)
                                in
-                               match p_assume st323 (hrmStoreBranchTime nodeLabel1 host1) x485 with (st324, x486)
+                               match p_assume st323 hrmStoreAssume x485 with (st324, x486)
                                in
                                match
                                  p_map
@@ -6705,7 +6678,7 @@
                     in
                     match p_apply st594 x928 x930 with (st595, x931)
                     in
-                    match p_assume st595 (hrmStoreBranchTime nodeLabel13 x922) x931 with (st596, x932)
+                    match p_assume st595 hrmStoreAssume x931 with (st596, x932)
                     in
                     match
                       p_map
@@ -6751,7 +6724,7 @@
                     let st601 =
                       p_weight
                         st600
-                        (hrmStoreBridgeSuppWeight nodeLabel13 x922)
+                        hrmStoreWeight
                         (lam x939.
                            externalLog
                              (addf
@@ -6840,7 +6813,7 @@
                     in
                     match p_apply st582 x908 x910 with (st583, x912)
                     in
-                    match p_assume st583 (hrmStoreBranchTime nodeLabel12 x902) x912 with (st584, x913)
+                    match p_assume st583 hrmStoreAssume x912 with (st584, x913)
                     in
                     match
                       p_map
@@ -6886,7 +6859,7 @@
                     let st589 =
                       p_weight
                         st588
-                        (hrmStoreBridgeSuppWeight nodeLabel12 x902)
+                        hrmStoreWeight
                         (lam x920.
                            externalLog
                              (addf
@@ -8557,22 +8530,10 @@
                         in
                         match p_apply st426 x673 x675 with (st427, x676)
                         in
-                        let nl = (match tree2 with MsgLeaf carried24
-                                             then
-                                               carried24.label
-                                             else match tree2 with MsgNode carried25
-                                             then
-                                               carried25.label
-                                             else
-                                               let #var"263" =
-                                                 print
-                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/flat-root-prior-HRM.tppl 129:90-129:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
-                                               in
-                                               exit 1) in
                         let st428 =
                           p_weight
                             st427
-                            (hrmStoreLikrWeight nl)
+                            hrmStoreWeight
                             (/-temp-/lam x679.
                                x679)
                             x676
@@ -8669,22 +8630,10 @@
                       in
                       match bool2real st432 x680 with (st433, x681)
                       in
-                      let nl = (match tree2 with MsgLeaf carried50
-                                         then
-                                           carried50.label
-                                         else match tree2 with MsgNode carried51
-                                         then
-                                           carried51.label
-                                         else
-                                           let #var"293" =
-                                             print
-                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/flat-root-prior-HRM.tppl 141:43-141:53>:\nField \'label\' not found\n[0m    weight bool2real(any(is2, rep)) + Real([31mtree.label[0m[0m) * 0.;\n"
-                                           in
-                                           exit 1) in
                       let st434 =
                         p_weight
                           st433
-                          (hrmStoreNodeSuppWeight nl)
+                          hrmStoreWeight
                           (lam x702.
                              externalLog
                                (addf
@@ -8818,22 +8767,10 @@
                       in
                       match p_apply st440 x688 x691 with (st441, x692)
                       in
-                      let nl = (match tree2 with MsgLeaf carried40
-                                         then
-                                           carried40.label
-                                         else match tree2 with MsgNode carried41
-                                         then
-                                           carried41.label
-                                         else
-                                           let #var"276" =
-                                             print
-                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 211:113-211:123>:\nField \'label\' not found\n[0m    logWeight branchSample.logModelDensity - branchSample.logSamplingDensity - nodeLogSamplingDensity * exp(Real([31mtree.label[0m[0m) * 0.);\n"
-                                           in
-                                           exit 1) in
                       let st442 =
                         p_weight
                           st441
-                          (hrmStoreLikrWeight nl)
+                          hrmStoreWeight
                           (/-temp-/lam x700.
                              x700)
                           x692
@@ -8841,10 +8778,6 @@
                       let foo2 = {} in
                       match observationMessage1 st442 rep2 with (st443, newMsg)
                       in
-                      -- Export node messages
-                      let st443 = p_export st443 (hrmStoreNodeMsg nl) newMsg in
-                      let st443 = p_export st443 (hrmStorePreorderMsg nl) preorderMsg in
-                      -- Done
                       match
                         mtxMul
                           st443
@@ -9090,27 +9023,19 @@
   in
   let subrootLength = #var"X11" in
   let nHosts1 = length11 hostDistances in
-  -- Export the input tree
-  recursive let buildTree = lam stree.
-    match stree with Leaf { label = label, age = _ } then 
-      HRMLeaf { label = label }
-    else match stree with Node { label = label, age = _, left = left, right = right } in 
-      HRMNode { label = label, left = buildTree left, right = buildTree right } in
-  match p_export st (hrmStoreTree (buildTree symbiontTree) (match input1 with {interactions = ints} in ints)) (p_pure ()) with st in
-  -- Done
   match
     p_assume
-      st hrmStoreLambda (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
+      st hrmStoreAssume (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
   with
     (st449, lambda)
   in
   match
-    p_assume st449 hrmStoreMu (p_pure (mkExponential 10.))
+    p_assume st449 hrmStoreAssume (p_pure (mkExponential 10.))
   with
     (st450, mu)
   in
   match
-    p_assume st450 hrmStoreBeta (p_pure (mkExponential 1.))
+    p_assume st450 hrmStoreAssume (p_pure (mkExponential 1.))
   with
     (st451, beta)
   in
@@ -9442,7 +9367,6 @@
   in
   match eitherEither st497 x776 with (st498, subrootSamplingProb)
   in
-  let st497 = p_export st497 (hrmStorePreorderMsg subrootLabel1) (p_pure b4) in
   match
     suggestNodeRep st498 subrootSamplingProb nHosts1 (negi 1)
   with
@@ -9455,7 +9379,7 @@
   let st502 =
     p_weight
       st501
-      (hrmStoreNodeSuppWeight subrootLabel1)
+      hrmStoreWeight
       (lam x805.
          externalLog (addf x805 (mulf (int2float subrootLabel1) 0.)))
       x778
@@ -9469,7 +9393,7 @@
   let st504 =
     p_weight
       st503
-      (hrmStoreLikrWeight subrootLabel1)
+      hrmStoreWeight
       (lam x804.
          addf
            (subf
@@ -9483,7 +9407,6 @@
   let foo7 = {} in
   match observationMessage1 st504 subrootRep with (st505, x780)
   in
-  let st505 = p_export st505 (hrmStoreNodeMsg subrootLabel1) x780 in
   match mtxMul st505 x780 subrootKernel with (st506, x781)
   in
   let subrootInfo =
@@ -9543,22 +9466,10 @@
   in
   match bool2real st509 x782 with (st510, x783)
   in
-  let rl = (match postorderTree with MsgLeaf carried56
-       then
-         carried56.label
-       else match postorderTree with MsgNode carried57
-       then
-         carried57.label
-       else
-         let #var"292" =
-           print
-             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 87:57-87:76>:\nField \'label\' not found\n[0m  let rootRep = suggestNodeRep(rootSamplingProb, nHosts, [31mpostorderTree.label[0m[0m);\n"
-         in
-         exit 1) in
   let st511 =
     p_weight
       st510
-      (hrmStoreNodeSuppWeight rl)
+      hrmStoreWeight
       (lam x803.
          externalLog
            (addf
@@ -9701,39 +9612,14 @@
   in
   match p_apply st517 x791 x794 with (st518, x795)
   in
-  let rl = (match postorderTree with MsgLeaf carried58
-                     then
-                       carried58.label
-                     else match postorderTree with MsgNode carried59
-                     then
-                       carried59.label
-                     else
-                       let #var"297" =
-                         print
-                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 110:40-110:59>:\nField \'label\' not found\n[0m    - rootLogSamplingDensity * exp(Real([31mpostorderTree.label[0m[0m) * 0.)\n"
-                       in
-                       exit 1) in
   let st519 =
     p_weight
-      st518 (hrmStoreLikrWeight rl) (/-temp-/lam x802.
+      st518 hrmStoreWeight (/-temp-/lam x802.
          x802) x795
   in
   let foo4 = {} in
   match observationMessage1 st519 rootRep with (st520, newMsg1)
   in
-  -- Export root messages
-  let st520 = p_export st520 (hrmStoreNodeMsg rl) newMsg1 in
-  let subrootMsg = (match subrootInfo with SubrootInfo1 x1091
-       then
-         x1091.msg
-       else
-         let #var"290" =
-           print
-             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 86:36-86:51>:\nField \'msg\' not found\n[0m  let rootSamplingProb = mtxElemMul([31msubrootInfo.msg[0m[0m, postorderTree.outMsg);\n"
-         in
-         p_pure (exit 1)) in
-  let st520 = p_export st520 (hrmStorePreorderMsg rl) subrootMsg in
-  -- Done
   let rootAge1 =
     match postorderTree with MsgLeaf carried64
     then
@@ -9855,48 +9741,44 @@
   with
     (st525, x801)
   in
-  -- Set up exports
-  let createExport = lam st. lam left. lam right.
-    let m = mapEmpty subi in
-    let makeDirtyArray = lam st. lam arr.
-      p_traverseSeq st (lam st. lam a. (st, a)) arr 
-    in
-    -- Traverse the tree to extract the node repertoires
-    recursive let extractSamples = lam st. lam m. lam hTree.
-      match hTree with HistoryLeaf _ then (st, m)
-      else match hTree with HistoryNode n in
-      match makeDirtyArray st n.repertoire with (st, wrappedRep) in
-      let m = mapInsert n.label wrappedRep m in
-      match extractSamples st m n.left with (st, m) in
-      extractSamples st m n.right
-    in
-    -- Save the root repertoire
-    match makeDirtyArray st rootRep with (st, dRootRep) in
-    let m = mapInsert rl dRootRep m in
-    -- Save the subroot repertoire
-    match makeDirtyArray st subrootRep with (st, dSubrootRep) in
-    let m = mapInsert subrootLabel1 dSubrootRep m in
-    -- Save the two subtrees
-    match extractSamples st m left with (st, m) in
-    match extractSamples st m right with (st, m) in
-    let dirtyKeyVal = lam st. lam kv. match kv with (k, v) in p_map st (lam v. (k, v)) v in 
-    match p_traverseSeq st dirtyKeyVal (mapToSeq m) with (st, wrappedReps) in
-    match p_map st (
-      lam mu. lam beta. lam lambda. lam reps.
-        -- printLn (join ["Export root rep: [", int2string rootRep1, ", ", int2string rootRep2, "]"]);
-        { mu = mu
-        , beta = beta
-        , lambda = lambda
-        , reps = mapFromSeq subi reps 
-        }
-    ) mu with (st, partres) in
-    match p_apply st partres beta with (st, partres) in
-    match p_apply st partres lambda with (st, partres) in
-    p_apply st partres wrappedReps
-  in
-  match createExport st525 x797 x799 with (st525, res) in
-  -- Done
   p_export
     st525
     hrmStoreExport
-    res
\ No newline at end of file
+    (p_pure
+       (dprint
+          (ReturnType1
+             { mu = mu,
+               beta = beta,
+               tree =
+                 HistoryNode
+                   { age =
+                       match symbiontTree with Leaf carried60
+                       then
+                         carried60.age
+                       else match symbiontTree with Node carried61
+                       then
+                         carried61.age
+                       else
+                         let #var"2105" =
+                           print
+                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 130:10-130:26>:\nField \'age\' not found\n[0m    age = [31msymbiontTree.age[0m[0m, label = symbiontTree.label,\n"
+                         in
+                         exit 1,
+                     label =
+                       match symbiontTree with Leaf carried62
+                       then
+                         carried62.label
+                       else match symbiontTree with Node carried63
+                       then
+                         carried63.label
+                       else
+                         let #var"2106" =
+                           print
+                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/subroot-HRM.tppl 130:36-130:54>:\nField \'label\' not found\n[0m    age = symbiontTree.age, label = [31msymbiontTree.label[0m[0m,\n"
+                         in
+                         exit 1,
+                     left = x797,
+                     right = x799,
+                     history = x801,
+                     repertoire = rootRep },
+               lambda = lambda })))
