--- build/transformed-model.mc	2026-01-26 10:43:25.712485855 +0100
+++ simple-HRM-tranformed.mc	2026-01-26 11:56:09.099161980 +0100
@@ -429,8 +429,8 @@
                 match
                   p_bind
                     st11
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st12.
                        lam x17.
                          match x17 with true
@@ -796,8 +796,8 @@
             match
               p_bind
                 st615
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st616.
                    lam x984.
                      match
@@ -855,8 +855,8 @@
             match
               p_bind
                 st605
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st606.
                    lam x975.
                      match
@@ -1189,8 +1189,8 @@
         match
           p_bind
             st597
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st598.
                lam x965.
                  match lti 0 x965 with true
@@ -1261,8 +1261,8 @@
           in
           p_bind
             st27
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st28.
                lam x40.
                  match x40 with true
@@ -2028,8 +2028,8 @@
               match
                 p_bind
                   st39
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st40.
                      lam x61.
                        match eqi x61 2 with true
@@ -2133,8 +2133,8 @@
                       in
                       p_bind
                         st45
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st46.
                            lam x70.
                              match
@@ -2156,8 +2156,8 @@
                                in
                                p_bind
                                  st47
-                                 simpleStoreSubmodel
-                                 (simpleInit {})
+                                 hrmStoreSubmodel
+                                 (hrmInit {})
                                  (lam st48.
                                     lam x72.
                                       match eqi x72 1 with true
@@ -2178,8 +2178,8 @@
                 lam nhosts3.
                   p_bind
                     st49
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st50.
                        lam x74.
                          match gti eventIndex11 x74 with true
@@ -2196,8 +2196,8 @@
                          in
                          p_bind
                            st51
-                           simpleStoreSubmodel
-                           (simpleInit {})
+                           hrmStoreSubmodel
+                           (hrmInit {})
                            (lam st52.
                               lam x77.
                                 match
@@ -2219,8 +2219,8 @@
                                   in
                                   p_bind
                                     st53
-                                    simpleStoreSubmodel
-                                    (simpleInit {})
+                                    hrmStoreSubmodel
+                                    (hrmInit {})
                                     (lam st54.
                                        lam x79.
                                          match eqi x79 1 with true
@@ -2545,8 +2545,8 @@
             in
             p_bind
               st533
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st534.
                  lam x871.
                    match eqi x871 0 with true
@@ -2769,8 +2769,8 @@
             in
             p_bind
               st512
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st513.
                  lam x842.
                    match eqi x842 0 with true
@@ -2981,8 +2981,8 @@
             in
             p_bind
               st486
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st487.
                  lam x802.
                    match eqi x802 0 with true
@@ -3205,8 +3205,8 @@
               match
                 p_bind
                   st63
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st64.
                      lam x95.
                        match eqi x95 2 with true
@@ -3302,8 +3302,8 @@
             match
               p_bind
                 st471
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st472.
                    lam x781.
                      match eqi x781 2 with true
@@ -3321,8 +3321,8 @@
             in
             p_bind
               st474
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st475.
                  lam x785.
                    match x785 with true
@@ -3401,8 +3401,8 @@
             match
               p_bind
                 st464
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st465.
                    lam x772.
                      match eqi x772 2 with true
@@ -3420,8 +3420,8 @@
             in
             p_bind
               st467
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st468.
                  lam x776.
                    match x776 with true
@@ -3489,8 +3489,8 @@
             match
               p_bind
                 st452
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st453.
                    lam x756.
                      match eqi x756 2 with true
@@ -3522,8 +3522,8 @@
             in
             p_bind
               st455
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st456.
                  lam x762.
                    match x762 with true
@@ -3586,8 +3586,8 @@
                   match
                     p_bind
                       st66
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st67.
                          lam x100.
                            match eqi x100 2 with true
@@ -3628,8 +3628,8 @@
                   match
                     p_bind
                       st71
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st72.
                          lam x108.
                            match eqi (get x108 (subi x107 1)) 2 with true
@@ -3693,8 +3693,8 @@
                   match
                     p_bind
                       st76
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st77.
                          lam x116.
                            match eqi x116 0 with true
@@ -3735,8 +3735,8 @@
                   match
                     p_bind
                       st81
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st82.
                          lam x124.
                            match eqi (get x124 (subi x123 1)) 0 with true
@@ -3793,8 +3793,8 @@
                 match
                   p_bind
                     st439
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st440.
                        lam x735.
                          match eqi x735 2 with true
@@ -3838,8 +3838,8 @@
                 match
                   p_bind
                     st444
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st445.
                        lam x745.
                          match eqi x745 0 with true
@@ -3880,8 +3880,8 @@
                 match
                   p_bind
                     st431
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st432.
                        lam x720.
                          match eqi (get x720 (subi x719 1)) 2 with true
@@ -4067,7 +4067,7 @@
                             (st93, x139)
                           in
                           match
-                            p_subApply st93 simpleStoreSubmodel (simpleInit {}) x137 x139
+                            p_subApply st93 hrmStoreSubmodel (hrmInit {}) x137 x139
                           with
                             (st94, x140)
                           in
@@ -4259,7 +4259,7 @@
                             (st115, x173)
                           in
                           match
-                            p_subApply st115 simpleStoreSubmodel (simpleInit {}) x171 x173
+                            p_subApply st115 hrmStoreSubmodel (hrmInit {}) x171 x173
                           with
                             (st116, x174)
                           in
@@ -4343,8 +4343,8 @@
                       in
                       p_bind
                         st129
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st130.
                            lam x195.
                              match gti nextIndex11 x195 with true
@@ -4514,8 +4514,8 @@
                       in
                       p_bind
                         st145
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st146.
                            lam x219.
                              match gti nextIndex12 x219 with true
@@ -4701,8 +4701,8 @@
                       in
                       p_bind
                         st164
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st165.
                            lam x251.
                              match gti nextIndex13 x251 with true
@@ -4872,8 +4872,8 @@
                       in
                       p_bind
                         st180
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st181.
                            lam x275.
                              match gti nextIndex14 x275 with true
@@ -5063,9 +5063,10 @@
     lam st430.
       lam msg1.
         lam x717.
+          lam nodeLabel.
           p_assume
             st430
-            simpleStoreAssume
+            (hrmStoreNode nodeLabel x717)
             (p_pure (mkCategorical (mtx3ToSeq msg1 x717)))
   in
   recursive
@@ -5073,14 +5074,15 @@
       lam st198.
         lam c25.
           lam s8.
+            lam nodeLabel.
             match s8 with [ _,
                 _ ] ++ _
             then
-              match anon15 st198 c25 (get s8 0) with (st199, x306)
+              match anon15 st198 c25 (get s8 0) nodeLabel with (st199, x306)
               in
               match splitAt s8 1 with {#label"1" = x307}
               in
-              match rec8 st199 c25 x307 with (st200, x308)
+              match rec8 st199 c25 x307 nodeLabel with (st200, x308)
               in
               (st200, cons x306 x308)
             else match s8 with [ e12 ]
@@ -5088,7 +5090,7 @@
               match splitAt s8 1 with {#label"1" = x309}
               in
               let slice11 = x309 in
-              match anon15 st198 c25 e12 with (st201, x310)
+              match anon15 st198 c25 e12 nodeLabel with (st201, x310)
               in
               (st201, [ x310 ])
             else match s8 with ""
@@ -5099,7 +5101,7 @@
     lam st429.
       lam msg.
         lam nHosts4.
-          lam nodeLabel5.
+          lam nodeLabel.
             let i111 = subi (addi (subi nHosts4 1) 1) 1 in
             rec8
               st429
@@ -5109,6 +5111,7 @@
                  rec34 (subi i111 1) (cons (anon14 i111) "")
                else
                  "")
+              nodeLabel
   in
   recursive
     let sampleHostHistoryRec =
@@ -5122,8 +5125,8 @@
                     lam embeddedQMatrix11.
                       p_bind
                         st202
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st203.
                            lam x313.
                              match ltf x313 finalAge2 with true
@@ -5159,15 +5162,13 @@
                              with
                                (st204, x316)
                              in
-                             match p_assume st204 simpleStoreAssume x316 with (st205, nextState11)
+                             match p_assume st204 (hrmStoreBranchState nodeLabel host) x316 with (st205, nextState11)
                              in
                              match
                                p_map
                                  st205
                                  (lam x317.
                                     mkExponential
-                                      (addf
-                                         (addf
                                             (get
                                                (match embeddedQMatrix11 with EmbeddedMarkovChainMatrix1 x391
                                                 then
@@ -5179,13 +5180,12 @@
                                                   in
                                                   exit 1)
                                                (subi (addi x317 1) 1))
-                                            (mulf (int2float host) 0.))
-                                         (mulf (int2float nodeLabel) 0.)))
+                                 )
                                  nextState11
                              with
                                (st206, x318)
                              in
-                             match p_assume st206 simpleStoreAssume x318 with (st207, x319)
+                             match p_assume st206 (hrmStoreBranchTime nodeLabel host) x318 with (st207, x319)
                              in
                              match
                                p_map
@@ -5272,8 +5272,8 @@
                     lam embeddedQMatrix12.
                       p_bind
                         st213
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st214.
                            lam x336.
                              match ltf x336 finalAge21 with true
@@ -5315,7 +5315,7 @@
                              with
                                (st216, x345)
                              in
-                             match p_assume st216 simpleStoreAssume x345 with (st217, nextState12)
+                             match p_assume st216 (hrmStoreBranchState nodeLabel1 host1) x345 with (st217, nextState12)
                              in
                              match
                                p_map
@@ -5341,7 +5341,7 @@
                              with
                                (st218, x347)
                              in
-                             match p_assume st218 simpleStoreAssume x347 with (st219, x348)
+                             match p_assume st218 (hrmStoreBranchTime nodeLabel1 host1) x347 with (st219, x348)
                              in
                              match
                                p_map
@@ -5465,8 +5465,6 @@
                         st421
                         (lam x705.
                            mkExponential
-                             (addf
-                                (addf
                                    (get
                                       (match embeddedQMatrix51 with EmbeddedMarkovChainMatrix1 x1082
                                        then
@@ -5478,13 +5476,12 @@
                                          in
                                          exit 1)
                                       (subi (addi x705 1) 1))
-                                   (mulf (int2float x704) 0.))
-                                (mulf (int2float nodeLabel12) 0.)))
+                                )
                         startState1
                     with
                       (st422, x706)
                     in
-                    match p_assume st422 simpleStoreAssume x706 with (st423, x707)
+                    match p_assume st422 (hrmStoreBranchTime nodeLabel12 x704) x706 with (st423, x707)
                     in
                     match
                       p_map
@@ -5530,12 +5527,11 @@
                     let st428 =
                       p_weight
                         st427
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel12 x704)
                         (lam x715.
                            externalLog
-                             (addf
-                                (addf x715 (mulf (int2float x704) 0.))
-                                (mulf (int2float nodeLabel12) 0.)))
+                            x715
+                        )
                         x713
                     in
                     let foo51 = {} in
@@ -5579,8 +5575,6 @@
                         st413
                         (lam x693.
                            mkExponential
-                             (addf
-                                (addf
                                    (get
                                       (match embeddedQMatrix5 with EmbeddedMarkovChainMatrix1 x1081
                                        then
@@ -5592,13 +5586,12 @@
                                          in
                                          exit 1)
                                       (subi (addi x693 1) 1))
-                                   (mulf (int2float x692) 0.))
-                                (mulf (int2float nodeLabel11) 0.)))
+                                )
                         startState
                     with
                       (st414, x694)
                     in
-                    match p_assume st414 simpleStoreAssume x694 with (st415, x695)
+                    match p_assume st414 (hrmStoreBranchTime nodeLabel11 x692) x694 with (st415, x695)
                     in
                     match
                       p_map
@@ -5644,12 +5637,10 @@
                     let st420 =
                       p_weight
                         st419
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel11 x692)
                         (lam x703.
                            externalLog
-                             (addf
-                                (addf x703 (mulf (int2float x692) 0.))
-                                (mulf (int2float nodeLabel11) 0.)))
+                            x703)
                         x700
                     in
                     let foo5 = {} in
@@ -6367,8 +6358,8 @@
                         match
                           p_bind
                             st272
-                            simpleStoreSubmodel
-                            (simpleInit {})
+                            hrmStoreSubmodel
+                            (hrmInit {})
                             (lam st273.
                                lam x460.
                                  match
@@ -6561,8 +6552,8 @@
                                      match
                                        p_bind
                                          st290
-                                         simpleStoreSubmodel
-                                         (simpleInit {})
+                                         hrmStoreSubmodel
+                                         (hrmInit {})
                                          (lam st291.
                                             lam x498.
                                               match gti 1 x498 with true
@@ -6671,7 +6662,7 @@
                                                 (st299, x508)
                                               in
                                               match
-                                                p_subApply st299 simpleStoreSubmodel (simpleInit {}) x506 x508
+                                                p_subApply st299 hrmStoreSubmodel (hrmInit {}) x506 x508
                                               with
                                                 (st300, x509)
                                               in
@@ -6792,7 +6783,6 @@
                             st315
                             (lam x539.
                                lam x540.
-                                 mulf
                                    (subf
                                       (match x539 with CorrectedBranchSample1 x4610
                                        then
@@ -6803,23 +6793,7 @@
                                              "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 135:15-135:43>:\nField \'logModelDensity\' not found\n[0m    logWeight ([31mbranchSample.logModelDensity[0m[0m - branchSample.logSamplingDensity) * exp(Real(tree.label) * 0.);\n"
                                          in
                                          exit 1)
-                                      x540)
-                                   (exp1
-                                      (mulf
-                                         (int2float
-                                            (match tree2 with MsgLeaf carried26
-                                             then
-                                               carried26.label
-                                             else match tree2 with MsgNode carried27
-                                             then
-                                               carried27.label
-                                             else
-                                               let #var"162" =
-                                                 print
-                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 135:90-135:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
-                                               in
-                                               exit 1))
-                                         0.)))
+                                      x540))
                             branchSample
                         with
                           (st316, x541)
@@ -6846,7 +6820,18 @@
                         let st319 =
                           p_weight
                             st318
-                            simpleStoreWeight
+                            (hrmStoreLikrWeight (match tree2 with MsgLeaf carried26
+                                             then
+                                               carried26.label
+                                             else match tree2 with MsgNode carried27
+                                             then
+                                               carried27.label
+                                             else
+                                               let #var"162" =
+                                                 print
+                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 135:90-135:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
+                                               in
+                                               exit 1))
                             (/-temp-/lam x547.
                                x547)
                             x544
@@ -6942,14 +6927,7 @@
                         let st324 =
                           p_weight
                             st323
-                            simpleStoreWeight
-                            (lam x655.
-                               externalLog
-                                 (addf
-                                    x655
-                                    (mulf
-                                       (int2float
-                                          (match tree2 with MsgLeaf carried50
+                            (hrmStoreBranchSuppWeight (match tree2 with MsgLeaf carried50
                                            then
                                              carried50.label
                                            else match tree2 with MsgNode carried51
@@ -6961,7 +6939,9 @@
                                                  "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 147:43-147:53>:\nField \'label\' not found\n[0m    weight bool2real(any(is2, rep)) + Real([31mtree.label[0m[0m) * 0.;\n"
                                              in
                                              exit 1))
-                                       0.)))
+                            (lam x655.
+                               externalLog
+                                    x655)
                             x550
                         in
                         let foo1 = {} in
@@ -7017,8 +6997,8 @@
                         match
                           p_bind
                             st329
-                            simpleStoreSubmodel
-                            (simpleInit {})
+                            hrmStoreSubmodel
+                            (hrmInit {})
                             (lam st330.
                                lam x553.
                                  match
@@ -7226,8 +7206,8 @@
                                      match
                                        p_bind
                                          st349
-                                         simpleStoreSubmodel
-                                         (simpleInit {})
+                                         hrmStoreSubmodel
+                                         (hrmInit {})
                                          (lam st350.
                                             lam x596.
                                               match gti 1 x596 with true
@@ -7336,7 +7316,7 @@
                                                 (st358, x606)
                                               in
                                               match
-                                                p_subApply st358 simpleStoreSubmodel (simpleInit {}) x604 x606
+                                                p_subApply st358 hrmStoreSubmodel (hrmInit {}) x604 x606
                                               with
                                                 (st359, x607)
                                               in
@@ -7502,13 +7482,17 @@
                         match
                           p_map
                             st378
-                            (lam x645.
-                               mulf
-                                 x645
-                                 (exp1
-                                    (mulf
-                                       (int2float
-                                          (match tree2 with MsgLeaf carried38
+                            (lam x645. x645)
+                            x644
+                        with
+                          (st379, x646)
+                        in
+                        match p_apply st379 x643 x646 with (st380, x647)
+                        in
+                        let st381 =
+                          p_weight
+                            st380
+                            (hrmStoreLikrWeight (match tree2 with MsgLeaf carried38
                                            then
                                              carried38.label
                                            else match tree2 with MsgNode carried39
@@ -7520,21 +7504,20 @@
                                                  "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 163:113-163:123>:\nField \'label\' not found\n[0m    logWeight branchSample.logModelDensity - branchSample.logSamplingDensity - nodeLogSamplingDensity * exp(Real([31mtree.label[0m[0m) * 0.);\n"
                                              in
                                              exit 1))
-                                       0.)))
-                            x644
-                        with
-                          (st379, x646)
-                        in
-                        match p_apply st379 x643 x646 with (st380, x647)
-                        in
-                        let st381 =
-                          p_weight
-                            st380
-                            simpleStoreWeight
                             (/-temp-/lam x654.
                                x654)
                             x647
                         in
+                      let fetchLabel = lam n. match n with MsgNode n then n.label else match n with MsgLeaf n in n.label in
+                      match p_export
+                        st381
+                        (hrmStoreLocTopo
+                          ( match tree2 with MsgNode n in n.label
+                          , match tree2 with MsgNode n in fetchLabel n.left
+                          , match tree2 with MsgNode n in fetchLabel n.right
+                          , false
+                          )
+                          ) (p_pure true) with st381 in
                         let foo2 = {} in
                         match observationMessage1 st381 rep2 1 nHosts with (st382, x648)
                         in
@@ -7590,6 +7573,7 @@
                         with
                           (st384, x650)
                         in
+
                         match
                           sampleTreeHistory
                             st384
@@ -7690,19 +7674,30 @@
                             repertoire = rep2 })
   in
   let input1 =
-    { dMean = 1.,
-      interactions = [ [ 1, 1 ],
-          [ 1, 1 ] ],
-      symbiontTree =
-        Node
-          { age = 1.,
-            label = 3,
-            left = Leaf
-                { age = 0., label = 1 },
-            right = Leaf
-                { age = 0., label = 2 } },
-      hostDistances = [ [ 0., 1. ],
-          [ 1., 0. ] ] }
+  { symbiontTree = Node
+    { age = 10.0
+    , label = 5
+    , left = Node
+      { label = 4
+      , age = 5.0
+      , left = Leaf
+        { label = 1
+        , age = 0.0
+        }
+      , right = Leaf
+        { label = 2
+        , age = 0.0
+        }
+      }
+    , right = Leaf
+      { label = 3
+      , age = 0.0
+      }
+    }
+  , interactions = [[2, 0], [1, 2], [2, 1]]
+  , hostDistances = [[1., 1.], [1., 1.]]
+  , dMean = 1.
+  }
   in
   match input1 with {symbiontTree = #var"X6"}
   in
@@ -7713,17 +7708,17 @@
   let nHosts1 = length11 hostDistances in
   match
     p_assume
-      st simpleStoreAssume (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
+      st hrmStoreLambda (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
   with
     (st387, lambda)
   in
   match
-    p_assume st387 simpleStoreAssume (p_pure (mkExponential 10.))
+    p_assume st387 hrmStoreMu (p_pure (mkExponential 10.))
   with
     (st388, mu)
   in
   match
-    p_assume st388 simpleStoreAssume (p_pure (mkExponential 1.))
+    p_assume st388 hrmStoreBeta (p_pure (mkExponential 1.))
   with
     (st389, beta)
   in
@@ -7879,14 +7874,7 @@
   let st404 =
     p_weight
       st403
-      simpleStoreWeight
-      (lam x685.
-         externalLog
-           (addf
-              x685
-              (mulf
-                 (int2float
-                    (match postorderTree with MsgLeaf carried64
+      (hrmStoreBranchSuppWeight (match postorderTree with MsgLeaf carried64
                      then
                        carried64.label
                      else match postorderTree with MsgNode carried65
@@ -7898,7 +7886,10 @@
                            "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 42:45-42:64>:\nField \'label\' not found\n[0m  weight bool2real(any(is2, rootRep)) + Real([31mpostorderTree.label[0m[0m) * 0.;\n"
                        in
                        exit 1))
-                 0.)))
+      (lam x685.
+         externalLog
+            x685
+      )
       x677
   in
   let foo3 = {} in
@@ -7910,30 +7901,20 @@
   let st406 =
     p_weight
       st405
-      simpleStoreWeight
+      (hrmStoreLikrWeight (match postorderTree with MsgLeaf carried62
+                     then
+                       carried62.label
+                     else match postorderTree with MsgNode carried63
+                    in
+                       carried63.label
+      ))
       (lam x684.
-         mulf
            (subf
               (negf
                  (log1
                     (subf (pow 3. (int2float nHosts1)) (pow 2. (int2float nHosts1)))))
               x684)
-           (exp1
-              (mulf
-                 (int2float
-                    (match postorderTree with MsgLeaf carried62
-                     then
-                       carried62.label
-                     else match postorderTree with MsgNode carried63
-                     then
-                       carried63.label
-                     else
-                       let #var"1103" =
-                         print
-                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 49:70-49:89>:\nField \'label\' not found\n[0m  logWeight (rootLogModelDensity - rootLogSamplingDensity) * exp(Real([31mpostorderTree.label[0m[0m) * 0.);\n"
-                       in
-                       exit 1))
-                 0.)))
+      )
       x678
   in
   let foo4 = {} in
@@ -7988,6 +7969,16 @@
       in
       exit 1
   in
+  let fetchLabel = lam n. match n with MsgNode n then n.label else match n with MsgLeaf n in n.label in
+  match p_export
+    st408
+    (hrmStoreLocTopo
+      ( match postorderTree with MsgNode n in n.label
+      , match postorderTree with MsgNode n in fetchLabel n.left
+      , match postorderTree with MsgNode n in fetchLabel n.right
+      , true
+      )
+    ) (p_pure true) with st408 in
   match
     sampleTreeHistory
       st408
@@ -8066,44 +8057,10 @@
   with
     (st410, x682)
   in
+  match p_map st410 (lam mu. lam beta. lam lambda. {mu = mu, beta = beta, lambda = lambda}) mu with (st, partres) in
+  match p_apply st partres beta with (st, partres) in
+  match p_apply st partres lambda with (st410, res) in
   p_export
     st410
-    simpleStoreExport
-    (p_pure
-       (dprint
-          (ReturnType1
-             { mu = mu,
-               beta = beta,
-               tree =
-                 HistoryNode
-                   { age =
-                       match symbiontTree with Leaf carried56
-                       then
-                         carried56.age
-                       else match symbiontTree with Node carried57
-                       then
-                         carried57.age
-                       else
-                         let #var"1100" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 76:10-76:26>:\nField \'age\' not found\n[0m    age = [31msymbiontTree.age[0m[0m, label = symbiontTree.label,\n"
-                         in
-                         exit 1,
-                     label =
-                       match symbiontTree with Leaf carried58
-                       then
-                         carried58.label
-                       else match symbiontTree with Node carried59
-                       then
-                         carried59.label
-                       else
-                         let #var"1101" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 76:36-76:54>:\nField \'label\' not found\n[0m    age = symbiontTree.age, label = [31msymbiontTree.label[0m[0m,\n"
-                         in
-                         exit 1,
-                     left = x680,
-                     right = x682,
-                     history = "",
-                     repertoire = rootRep },
-               lambda = lambda })))
+    hrmStoreExport
+    res 
\ No newline at end of file
