--- build/transformed-model.mc	2026-02-26 15:19:17.411815631 +0100
+++ simple-input-HRM.mc	2026-02-26 19:39:40.762471053 +0100
@@ -1,3 +1,23 @@
+  type TreeLabeled p18 p19 p20 in
+  con Leaf: all p21. all p22. all p23. p21 -> TreeLabeled p21 p22 p23
+  in
+  con Node: all p24. all p25. all p26. {age: p25, left: TreeLabeled p24 p25 p26, label: p26, right: TreeLabeled p24 p25 p26} -> TreeLabeled p24 p25 p26
+  in
+let input1 =
+  { interactions = [ [ 2, 1 ],
+        [ 1, 2 ] ],
+    symbiontTree =
+      Node
+        { age = 1.,
+          label = 3,
+          left = Leaf
+              { age = 0., label = 1 },
+          right = Leaf
+              { age = 0., label = 2 } },
+    hostDistances = [ [ 0., 1. ],
+        [ 1., 0. ] ] }
+in
+
 lam st.
   type Option p p1 in
   con Some: all p2. all p3. p2 -> Option p2 p3 in
@@ -13,11 +33,7 @@
   con DimensionMismatch: all p14. all p15. p14 -> MatError p14 p15
   in
   con NotSquare: all p16. all p17. p17 -> MatError p16 p17 in
-  type TreeLabeled p18 p19 p20 in
-  con Leaf: all p21. all p22. all p23. p21 -> TreeLabeled p21 p22 p23
-  in
-  con Node: all p24. all p25. all p26. {age: p25, left: TreeLabeled p24 p25 p26, label: p26, right: TreeLabeled p24 p25 p26} -> TreeLabeled p24 p25 p26
-  in
+
   type MsgTree p27 p28 p29 p30 p31 p32 p33 p34 in
   con MsgLeaf: all p35. all p36. all p37. all p38. all p39. all p40. all p41. all p42. p35 -> MsgTree p35 p36 p37 p38 p39 p40 p41 p42
   in
@@ -497,7 +513,7 @@
             (st921, x1474)
           in
           match
-            p_subApply st921 simpleStoreSubmodel (simpleInit {}) x1470 x1474
+            p_subApply st921 hrmStoreSubmodel (hrmInit {}) x1470 x1474
           with
             (st922, x1475)
           in
@@ -700,8 +716,8 @@
                 match
                   p_bind
                     st9
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st10.
                        lam x10.
                          match x10 with true
@@ -1170,8 +1186,8 @@
             match
               p_bind
                 st851
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st852.
                    lam x1318.
                      match
@@ -1229,8 +1245,8 @@
             match
               p_bind
                 st841
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st842.
                    lam x1306.
                      match
@@ -1530,7 +1546,7 @@
                 (st811, x1264)
               in
               match
-                p_subApply st811 simpleStoreSubmodel (simpleInit {}) x1262 x1264
+                p_subApply st811 hrmStoreSubmodel (hrmInit {}) x1262 x1264
               with
                 (st812, x1265)
               in
@@ -1567,8 +1583,8 @@
               match
                 p_bind
                   st817
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st818.
                      lam x1273.
                        match
@@ -1638,8 +1654,8 @@
             in
             p_subMap
               st795
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam x1228.
                  lam st796.
                    match geqi 0 x1228 with true
@@ -1741,8 +1757,8 @@
         match
           p_subMap
             st785
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam x1221.
                lam st786.
                  match geqi 0 x1221 with true
@@ -1767,8 +1783,8 @@
         match
           p_subMap
             st790
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam x1224.
                lam st791.
                  match geqi 0 x1224 with true
@@ -1852,8 +1868,8 @@
           match
             p_bind
               st769
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st770.
                  lam x1194.
                    match
@@ -1979,8 +1995,8 @@
         match
           p_bind
             st755
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st756.
                lam x1171.
                  match lti 0 x1171 with true
@@ -2220,6 +2236,19 @@
                          1
                          nhosts1)
                   in
+                  let ll = (match tree1 with Leaf carried4
+                          then
+                            carried4.label
+                          else match tree1 with Node carried5
+                          then
+                            carried5.label
+                          else
+                            let #var"22" =
+                              print
+                                "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/host-rep-lib/belief-propagation.tppl 10:32-10:42>:\nField \'label\' not found\n[0m    let leafInts = interactions[[31mtree.label[0m[0m];\n"
+                            in
+                            exit 1) in
+
                   let leafInts =
                     get
                       interactions1
@@ -2238,6 +2267,8 @@
                             exit 1)
                          1)
                   in
+                  let __outMsg = mtxCreate nhosts1 3 (observationMessage leafInts 1 nhosts1) in
+                  let st36 = p_export st36 (hrmStoreNodeMsg ll) (p_pure __outMsg) in
                   (st36, MsgLeaf
                     { age = 0.,
                       label =
@@ -2362,6 +2393,12 @@
                 in
                 match mtxTrans st42 leftKernel with (st43, x43)
                 in
+                -- Save the transition kernels
+                let getLabel = lam n. match n with MsgLeaf l then l.label else match n with MsgNode n in n.label in
+                let leftLabel = getLabel left in
+                let rightLabel = getLabel right in
+                let st43 = p_export st43 (hrmStoreTransKernel rightLabel) rightKernel in 
+                let st43 = p_export st43 (hrmStoreTransKernel leftLabel) leftKernel in 
                 match
                   mtxMul
                     st43
@@ -2533,8 +2570,8 @@
                   match
                     p_bind
                       st50
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st51.
                          lam x50.
                            match
@@ -2595,8 +2632,8 @@
               match
                 p_bind
                   st737
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st738.
                      lam x1144.
                        match
@@ -2658,8 +2695,8 @@
               match
                 p_bind
                   st54
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st55.
                      lam x55.
                        match eqi x55 2 with true
@@ -2995,8 +3032,8 @@
                       in
                       p_bind
                         st58
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st59.
                            lam x61.
                              match
@@ -3018,8 +3055,8 @@
                                in
                                p_bind
                                  st60
-                                 simpleStoreSubmodel
-                                 (simpleInit {})
+                                 hrmStoreSubmodel
+                                 (hrmInit {})
                                  (lam st61.
                                     lam x63.
                                       match eqi x63 1 with true
@@ -3040,8 +3077,8 @@
                 lam nhosts3.
                   p_bind
                     st62
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st63.
                        lam x65.
                          match gti eventIndex11 x65 with true
@@ -3058,8 +3095,8 @@
                          in
                          p_bind
                            st64
-                           simpleStoreSubmodel
-                           (simpleInit {})
+                           hrmStoreSubmodel
+                           (hrmInit {})
                            (lam st65.
                               lam x68.
                                 match
@@ -3081,8 +3118,8 @@
                                   in
                                   p_bind
                                     st66
-                                    simpleStoreSubmodel
-                                    (simpleInit {})
+                                    hrmStoreSubmodel
+                                    (hrmInit {})
                                     (lam st67.
                                        lam x70.
                                          match eqi x70 1 with true
@@ -3406,8 +3443,8 @@
             in
             p_bind
               st674
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st675.
                  lam x1049.
                    match eqi x1049 0 with true
@@ -3596,8 +3633,8 @@
             in
             p_bind
               st655
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st656.
                  lam x1024.
                    match eqi x1024 0 with true
@@ -3614,8 +3651,16 @@
                      in
                      match sapply st657 repertoire12 with (st658, x1028)
                      in
+                     let pf = lam s. printLn (join ["Repertoire [", strJoin ", " (map int2string s), "]"]) in
+                     let makeDirtyArray = lam st. lam arr.
+                       p_traverseSeq st (lam st. lam a. (st, a)) arr 
+                     in
+                     match makeDirtyArray st657 repertoire12 with(st657, tmp) in
+                     match p_map st657 pf tmp with (st657, _) in
                      match whichTrue1 st658 x1028 with (st659, x1029)
                      in
+                     let pf = lam s. printLn (join ["Which true [", strJoin ", " (map int2string s), "]"]) in
+                     match p_map st658 pf x1029 with (st658, _) in
                      match
                        mtxRowCols2
                          st659
@@ -3635,6 +3680,8 @@
                      in
                      match mtxMean1 st660 x1030 with (st661, x1031)
                      in
+                     let pf = lam s. printLn (join ["Normalized distance ", float2string s]) in
+                     match p_map st661 pf x1031 with (st661, _) in
                      match
                        p_map
                          st661
@@ -3770,8 +3817,8 @@
             in
             p_bind
               st631
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st632.
                  lam x988.
                    match eqi x988 0 with true
@@ -3977,8 +4024,8 @@
             match
               p_bind
                 st617
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st618.
                    lam x972.
                      match eqi x972 2 with true
@@ -3996,8 +4043,8 @@
             in
             p_bind
               st620
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st621.
                  lam x976.
                    match x976 with true
@@ -4076,8 +4123,8 @@
             match
               p_bind
                 st610
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st611.
                    lam x963.
                      match eqi x963 2 with true
@@ -4095,8 +4142,8 @@
             in
             p_bind
               st613
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st614.
                  lam x967.
                    match x967 with true
@@ -4164,8 +4211,8 @@
             match
               p_bind
                 st598
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st599.
                    lam x946.
                      match eqi x946 2 with true
@@ -4197,8 +4244,8 @@
             in
             p_bind
               st601
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st602.
                  lam x953.
                    match x953 with true
@@ -4261,8 +4308,8 @@
                   match
                     p_bind
                       st76
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st77.
                          lam x86.
                            match eqi x86 2 with true
@@ -4303,8 +4350,8 @@
                   match
                     p_bind
                       st81
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st82.
                          lam x94.
                            match eqi (get x94 (subi x93 1)) 2 with true
@@ -4368,8 +4415,8 @@
                   match
                     p_bind
                       st86
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st87.
                          lam x102.
                            match eqi x102 0 with true
@@ -4410,8 +4457,8 @@
                   match
                     p_bind
                       st91
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st92.
                          lam x110.
                            match eqi (get x110 (subi x109 1)) 0 with true
@@ -4475,8 +4522,8 @@
                 match
                   p_bind
                     st585
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st586.
                        lam x924.
                          match eqi x924 2 with true
@@ -4520,8 +4567,8 @@
                 match
                   p_bind
                     st590
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st591.
                        lam x934.
                          match eqi x934 0 with true
@@ -4562,8 +4609,8 @@
                 match
                   p_bind
                     st575
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st576.
                        lam x906.
                          match eqi (get x906 (subi x905 1)) 2 with true
@@ -4607,8 +4654,8 @@
                 match
                   p_bind
                     st580
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st581.
                        lam x917.
                          match eqi (get x917 (subi x916 1)) 0 with true
@@ -4757,7 +4804,7 @@
                             (st105, x128)
                           in
                           match
-                            p_subApply st105 simpleStoreSubmodel (simpleInit {}) x126 x128
+                            p_subApply st105 hrmStoreSubmodel (hrmInit {}) x126 x128
                           with
                             (st106, x129)
                           in
@@ -4949,7 +4996,7 @@
                             (st127, x165)
                           in
                           match
-                            p_subApply st127 simpleStoreSubmodel (simpleInit {}) x163 x165
+                            p_subApply st127 hrmStoreSubmodel (hrmInit {}) x163 x165
                           with
                             (st128, x166)
                           in
@@ -5033,8 +5080,8 @@
                       in
                       p_bind
                         st141
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st142.
                            lam x187.
                              match gti nextIndex11 x187 with true
@@ -5273,8 +5320,8 @@
                       in
                       p_bind
                         st170
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st171.
                            lam x234.
                              match gti nextIndex12 x234 with true
@@ -5520,8 +5567,8 @@
                       in
                       p_bind
                         st202
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st203.
                            lam x283.
                              match gti nextIndex13 x283 with true
@@ -5760,8 +5807,8 @@
                       in
                       p_bind
                         st231
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st232.
                            lam x338.
                              match gti nextIndex14 x338 with true
@@ -6033,7 +6080,7 @@
             with
               (st574, x903)
             in
-            p_assume st574 simpleStoreAssume x903
+            p_assume st574 (hrmStoreNode nodeLabel51 x897) x903
   in
   recursive
     let rec9 =
@@ -6091,8 +6138,8 @@
                     lam embeddedQMatrix11.
                       p_bind
                         st266
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st267.
                            lam x402.
                              match
@@ -6169,7 +6216,7 @@
                                in
                                match p_apply st275 x415 x417 with (st276, x418)
                                in
-                               match p_assume st276 simpleStoreAssume x418 with (st277, nextState11)
+                               match p_assume st276 (hrmStoreBranchState nodeLabel host) x418 with (st277, nextState11)
                                in
                                match
                                  p_traverseSeq
@@ -6216,7 +6263,7 @@
                                in
                                match p_apply st281 x424 x426 with (st282, x427)
                                in
-                               match p_assume st282 simpleStoreAssume x427 with (st283, x428)
+                               match p_assume st282 (hrmStoreBranchTime nodeLabel host) x427 with (st283, x428)
                                in
                                match
                                  p_map
@@ -6321,8 +6368,8 @@
                     lam embeddedQMatrix12.
                       p_bind
                         st291
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st292.
                            lam x443.
                              match
@@ -6402,7 +6449,7 @@
                                in
                                match p_apply st301 x457 x459 with (st302, x460)
                                in
-                               match p_assume st302 simpleStoreAssume x460 with (st303, nextState12)
+                               match p_assume st302 (hrmStoreBranchState nodeLabel1 host1) x460 with (st303, nextState12)
                                in
                                match
                                  p_traverseSeq
@@ -6449,7 +6496,7 @@
                                in
                                match p_apply st307 x466 x468 with (st308, x469)
                                in
-                               match p_assume st308 simpleStoreAssume x469 with (st309, x470)
+                               match p_assume st308 (hrmStoreBranchTime nodeLabel1 host1) x469 with (st309, x470)
                                in
                                match
                                  p_map
@@ -6631,7 +6678,7 @@
                     in
                     match p_apply st561 x883 x885 with (st562, x886)
                     in
-                    match p_assume st562 simpleStoreAssume x886 with (st563, x887)
+                    match p_assume st562 (hrmStoreBranchTime nodeLabel12 x876) x886 with (st563, x887)
                     in
                     match
                       p_map
@@ -6677,7 +6724,7 @@
                     let st568 =
                       p_weight
                         st567
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel12 x876)
                         (lam x895.
                            externalLog
                              (addf
@@ -6766,7 +6813,7 @@
                     in
                     match p_apply st549 x864 x866 with (st550, x867)
                     in
-                    match p_assume st550 simpleStoreAssume x867 with (st551, x868)
+                    match p_assume st550 (hrmStoreBranchTime nodeLabel11 x858) x867 with (st551, x868)
                     in
                     match
                       p_map
@@ -6812,7 +6859,7 @@
                     let st556 =
                       p_weight
                         st555
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel11 x858)
                         (lam x875.
                            externalLog
                              (addf
@@ -7468,8 +7515,8 @@
           in
           p_bind
             st360
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st361.
                lam x572.
                  match x572 with true
@@ -7564,8 +7611,8 @@
                         match
                           p_bind
                             st367
-                            simpleStoreSubmodel
-                            (simpleInit {})
+                            hrmStoreSubmodel
+                            (hrmInit {})
                             (lam st368.
                                lam x575.
                                  match
@@ -7758,8 +7805,8 @@
                                      match
                                        p_bind
                                          st385
-                                         simpleStoreSubmodel
-                                         (simpleInit {})
+                                         hrmStoreSubmodel
+                                         (hrmInit {})
                                          (lam st386.
                                             lam x613.
                                               match gti 1 x613 with true
@@ -7868,7 +7915,7 @@
                                                 (st394, x623)
                                               in
                                               match
-                                                p_subApply st394 simpleStoreSubmodel (simpleInit {}) x621 x623
+                                                p_subApply st394 hrmStoreSubmodel (hrmInit {}) x621 x623
                                               with
                                                 (st395, x624)
                                               in
@@ -8040,10 +8087,22 @@
                         in
                         match p_apply st412 x656 x658 with (st413, x659)
                         in
+                        let nl = (match tree2 with MsgLeaf carried26
+                                             then
+                                               carried26.label
+                                             else match tree2 with MsgNode carried27
+                                             then
+                                               carried27.label
+                                             else
+                                               let #var"263" =
+                                                 print
+                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 127:90-127:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
+                                               in
+                                               exit 1) in
                         let st414 =
                           p_weight
                             st413
-                            simpleStoreWeight
+                            (hrmStoreLikrWeight nl)
                             (/-temp-/lam x662.
                                x662)
                             x659
@@ -8140,10 +8199,22 @@
                       in
                       match bool2real st418 x663 with (st419, x664)
                       in
+                      let nl = (match tree2 with MsgLeaf carried52
+                                         then
+                                           carried52.label
+                                         else match tree2 with MsgNode carried53
+                                         then
+                                           carried53.label
+                                         else
+                                           let #var"293" =
+                                             print
+                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 139:43-139:53>:\nField \'label\' not found\n[0m    weight bool2real(any(is2, rep)) + Real([31mtree.label[0m[0m) * 0.;\n"
+                                           in
+                                           exit 1) in
                       let st420 =
                         p_weight
                           st419
-                          simpleStoreWeight
+                          (hrmStoreNodeSuppWeight nl)
                           (lam x770.
                              externalLog
                                (addf
@@ -8218,8 +8289,8 @@
                       match
                         p_bind
                           st425
-                          simpleStoreSubmodel
-                          (simpleInit {})
+                          hrmStoreSubmodel
+                          (hrmInit {})
                           (lam st426.
                              lam x667.
                                match
@@ -8427,8 +8498,8 @@
                                    match
                                      p_bind
                                        st445
-                                       simpleStoreSubmodel
-                                       (simpleInit {})
+                                       hrmStoreSubmodel
+                                       (hrmInit {})
                                        (lam st446.
                                           lam x710.
                                             match gti 1 x710 with true
@@ -8537,7 +8608,7 @@
                                               (st454, x720)
                                             in
                                             match
-                                              p_subApply st454 simpleStoreSubmodel (simpleInit {}) x718 x720
+                                              p_subApply st454 hrmStoreSubmodel (hrmInit {}) x718 x720
                                             with
                                               (st455, x721)
                                             in
@@ -8728,10 +8799,22 @@
                       in
                       match p_apply st475 x757 x760 with (st476, x761)
                       in
+                      let nl = (match tree2 with MsgLeaf carried40
+                                         then
+                                           carried40.label
+                                         else match tree2 with MsgNode carried41
+                                         then
+                                           carried41.label
+                                         else
+                                           let #var"280" =
+                                             print
+                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 155:113-155:123>:\nField \'label\' not found\n[0m    logWeight branchSample.logModelDensity - branchSample.logSamplingDensity - nodeLogSamplingDensity * exp(Real([31mtree.label[0m[0m) * 0.);\n"
+                                           in
+                                           exit 1) in
                       let st477 =
                         p_weight
                           st476
-                          simpleStoreWeight
+                          (hrmStoreLikrWeight nl)
                           (/-temp-/lam x769.
                              x769)
                           x761
@@ -8739,7 +8822,9 @@
                       let foo2 = {} in
                       match observationMessage1 st477 rep2 1 nHosts with (st478, x762)
                       in
-                      match mtxCreate1 st478 nHosts 3 x762 with (st479, newMsg)
+                      match mtxCreate1 st478 nHosts 3 x762 with (st479, newMsg) in
+                      let st479 = p_export st479 (hrmStoreNodeMsg nl) newMsg in
+                      let st479 = p_export st479 (hrmStorePreorderMsg nl) preorderMsg
                       in
                       match
                         mtxMul
@@ -8973,20 +9058,6 @@
             in
             ""
   in
-  let input1 =
-    { interactions = [ [ 2, 1 ],
-          [ 1, 2 ] ],
-      symbiontTree =
-        Node
-          { age = 1.,
-            label = 3,
-            left = Leaf
-                { age = 0., label = 1 },
-            right = Leaf
-                { age = 0., label = 2 } },
-      hostDistances = [ [ 0., 1. ],
-          [ 1., 0. ] ] }
-  in
   match input1 with {symbiontTree = #var"X9"}
   in
   let symbiontTree = #var"X9" in
@@ -8996,20 +9067,26 @@
   let nHosts1 = length1 hostDistances in
   match
     p_assume
-      st simpleStoreAssume (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
+      st hrmStoreLambda (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
   with
     (st485, lambda)
   in
   match
-    p_assume st485 simpleStoreAssume (p_pure (mkExponential 10.))
+    p_assume st485 hrmStoreMu (p_pure (mkExponential 10.))
   with
     (st486, mu)
   in
   match
-    p_assume st486 simpleStoreAssume (p_pure (mkExponential 1.))
+    p_assume st486 hrmStoreBeta (p_pure (mkExponential 1.))
   with
     (st487, beta)
   in
+  recursive let buildTree = lam stree.
+    match stree with Leaf { label = label, age = _ } then 
+      HRMLeaf { label = label }
+    else match stree with Node { label = label, age = _, left = left, right = right } in 
+      HRMNode { label = label, left = buildTree left, right = buildTree right } in
+  match p_export st487 (hrmStoreTree (buildTree symbiontTree) (match input1 with {interactions = ints} in ints)) (p_pure ()) with st487 in
   match
     p_map
       st487
@@ -9185,8 +9262,8 @@
   match
     p_bind
       st510
-      simpleStoreSubmodel
-      (simpleInit {})
+      hrmStoreSubmodel
+      (hrmInit {})
       (lam st511.
          lam x814.
            match
@@ -9255,10 +9332,22 @@
   in
   match bool2real st517 x821 with (st518, x822)
   in
+  let rl = (match postorderTree with MsgLeaf carried64
+                     then
+                       carried64.label
+                     else match postorderTree with MsgNode carried65
+                     then
+                       carried65.label
+                     else
+                       let #var"2105" =
+                         print
+                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 41:45-41:64>:\nField \'label\' not found\n[0m  weight bool2real(any(is2, rootRep)) + Real([31mpostorderTree.label[0m[0m) * 0.;\n"
+                       in
+                       exit 1) in
   let st519 =
     p_weight
       st518
-      simpleStoreWeight
+      (hrmStoreNodeSuppWeight rl)
       (lam x850.
          externalLog
            (addf
@@ -9286,10 +9375,22 @@
   with
     (st520, x823)
   in
+  let rl = (match postorderTree with MsgLeaf carried64
+                     then
+                       carried64.label
+                     else match postorderTree with MsgNode carried65
+                     then
+                       carried65.label
+                     else
+                       let #var"2105" =
+                         print
+                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 41:45-41:64>:\nField \'label\' not found\n[0m  weight bool2real(any(is2, rootRep)) + Real([31mpostorderTree.label[0m[0m) * 0.;\n"
+                       in
+                       exit 1) in
   let st521 =
     p_weight
       st520
-      simpleStoreWeight
+      (hrmStoreLikrWeight rl)
       (lam x849.
          mulf
            (subf
@@ -9318,7 +9419,9 @@
   let foo4 = {} in
   match observationMessage1 st521 rootRep 1 nHosts1 with (st522, x824)
   in
-  match mtxCreate1 st522 nHosts1 3 x824 with (st523, newMsg1)
+  match mtxCreate1 st522 nHosts1 3 x824 with (st523, newMsg1) in
+  let st523 = p_export st523 (hrmStoreNodeMsg rl) newMsg1 in
+  let st523 = p_export st523 (hrmStorePreorderMsg rl) (p_pure b4) 
   in
   let ns = rec15 (length1 hostDistances) hostDistances hostDistances
   in
@@ -9502,44 +9605,37 @@
   with
     (st541, x848)
   in
-  p_export
+  let m = mapEmpty subi in
+  let makeDirtyArray = lam st. lam arr.
+    p_traverseSeq st (lam st. lam a. (st, a)) arr 
+  in
+  recursive let extractSamples = lam st. lam m. lam hTree.
+    match hTree with HistoryLeaf _ then (st, m)
+    else match hTree with HistoryNode n in
+    match makeDirtyArray st n.repertoire with (st, wrappedRep) in
+    let m = mapInsert n.label wrappedRep m in
+    match extractSamples st m n.left with (st, m) in
+    extractSamples st m n.right
+  in
+  match makeDirtyArray st541 rootRep with (st541, dRootRep) in
+  let m = mapInsert rl dRootRep m in
+  match extractSamples st541 m x844 with (st541, m) in
+  match extractSamples st541 m x846 with (st541, m) in
+  let dirtyKeyVal = lam st. lam kv. match kv with (k, v) in p_map st (lam v. (k, v)) v in 
+  match p_traverseSeq st541 dirtyKeyVal (mapToSeq m) with (st541, wrappedReps) in
+  match p_map st541 (
+    lam mu. lam beta. lam lambda. lam reps.
+      -- printLn (join ["Export root rep: [", int2string rootRep1, ", ", int2string rootRep2, "]"]);
+      { mu = mu
+      , beta = beta
+      , lambda = lambda
+      , reps = mapFromSeq subi reps 
+      }
+  ) mu with (st541, partres) in
+  match p_apply st541 partres beta with (st541, partres) in
+  match p_apply st541 partres lambda with (st541, partres) in
+  match p_apply st541 partres wrappedReps with (st541, res) in
+   p_export
     st541
-    simpleStoreExport
-    (p_pure
-       (dprint
-          (ReturnType1
-             { mu = mu,
-               beta = beta,
-               tree =
-                 HistoryNode
-                   { age =
-                       match symbiontTree with Leaf carried56
-                       then
-                         carried56.age
-                       else match symbiontTree with Node carried57
-                       then
-                         carried57.age
-                       else
-                         let #var"2101" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 74:10-74:26>:\nField \'age\' not found\n[0m    age = [31msymbiontTree.age[0m[0m, label = symbiontTree.label,\n"
-                         in
-                         exit 1,
-                     label =
-                       match symbiontTree with Leaf carried58
-                       then
-                         carried58.label
-                       else match symbiontTree with Node carried59
-                       then
-                         carried59.label
-                       else
-                         let #var"2102" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 74:36-74:54>:\nField \'label\' not found\n[0m    age = symbiontTree.age, label = [31msymbiontTree.label[0m[0m,\n"
-                         in
-                         exit 1,
-                     left = x844,
-                     right = x846,
-                     history = x848,
-                     repertoire = rootRep },
-               lambda = lambda })))
+    hrmStoreExport
+    res
\ No newline at end of file
