--- build/transformed-model.mc	2026-02-22 17:36:06.633864588 +0100
+++ after-debug-hrm.mc	2026-02-21 17:42:42.812000176 +0100
@@ -426,8 +426,8 @@
                 match
                   p_bind
                     st7
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st8.
                        lam x8.
                          match x8 with true
@@ -937,8 +937,8 @@
             match
               p_bind
                 st817
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st818.
                    lam x1271.
                      match
@@ -996,8 +996,8 @@
             match
               p_bind
                 st807
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st808.
                    lam x1262.
                      match
@@ -1254,7 +1254,7 @@
             (st789, x1236)
           in
           match
-            p_subApply st789 simpleStoreSubmodel (simpleInit {}) x1234 x1236
+            p_subApply st789 hrmStoreSubmodel (hrmInit {}) x1234 x1236
           with
             (st790, x1237)
           in
@@ -1334,8 +1334,8 @@
           match
             p_bind
               st771
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st772.
                  lam x1205.
                    match
@@ -1468,8 +1468,8 @@
         match
           p_bind
             st758
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st759.
                lam x1179.
                  match lti 0 x1179 with true
@@ -1540,8 +1540,8 @@
           in
           p_bind
             st23
-            simpleStoreSubmodel
-            (simpleInit {})
+            hrmStoreSubmodel
+            (hrmInit {})
             (lam st24.
                lam x31.
                  match x31 with true
@@ -2306,8 +2306,8 @@
                   match
                     p_bind
                       st45
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st46.
                          lam x53.
                            match
@@ -2368,8 +2368,8 @@
               match
                 p_bind
                   st707
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st708.
                      lam x1111.
                        match
@@ -2411,8 +2411,8 @@
               match
                 p_bind
                   st49
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st50.
                      lam x57.
                        match eqi x57 2 with true
@@ -2516,8 +2516,8 @@
                       in
                       p_bind
                         st55
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st56.
                            lam x66.
                              match
@@ -2539,8 +2539,8 @@
                                in
                                p_bind
                                  st57
-                                 simpleStoreSubmodel
-                                 (simpleInit {})
+                                 hrmStoreSubmodel
+                                 (hrmInit {})
                                  (lam st58.
                                     lam x68.
                                       match eqi x68 1 with true
@@ -2561,8 +2561,8 @@
                 lam nhosts3.
                   p_bind
                     st59
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st60.
                        lam x70.
                          match gti eventIndex11 x70 with true
@@ -2579,8 +2579,8 @@
                          in
                          p_bind
                            st61
-                           simpleStoreSubmodel
-                           (simpleInit {})
+                           hrmStoreSubmodel
+                           (hrmInit {})
                            (lam st62.
                               lam x73.
                                 match
@@ -2602,8 +2602,8 @@
                                   in
                                   p_bind
                                     st63
-                                    simpleStoreSubmodel
-                                    (simpleInit {})
+                                    hrmStoreSubmodel
+                                    (hrmInit {})
                                     (lam st64.
                                        lam x75.
                                          match eqi x75 1 with true
@@ -2947,8 +2947,8 @@
             in
             p_bind
               st676
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st677.
                  lam x1060.
                    match eqi x1060 0 with true
@@ -3175,8 +3175,8 @@
             in
             p_bind
               st655
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st656.
                  lam x1029.
                    match eqi x1029 0 with true
@@ -3387,8 +3387,8 @@
             in
             p_bind
               st629
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st630.
                  lam x989.
                    match eqi x989 0 with true
@@ -3611,8 +3611,8 @@
               match
                 p_bind
                   st73
-                  simpleStoreSubmodel
-                  (simpleInit {})
+                  hrmStoreSubmodel
+                  (hrmInit {})
                   (lam st74.
                      lam x91.
                        match eqi x91 2 with true
@@ -3710,8 +3710,8 @@
             match
               p_bind
                 st613
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st614.
                    lam x968.
                      match eqi x968 2 with true
@@ -3729,8 +3729,8 @@
             in
             p_bind
               st616
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st617.
                  lam x972.
                    match x972 with true
@@ -3809,8 +3809,8 @@
             match
               p_bind
                 st606
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st607.
                    lam x959.
                      match eqi x959 2 with true
@@ -3828,8 +3828,8 @@
             in
             p_bind
               st609
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st610.
                  lam x963.
                    match x963 with true
@@ -3897,8 +3897,8 @@
             match
               p_bind
                 st594
-                simpleStoreSubmodel
-                (simpleInit {})
+                hrmStoreSubmodel
+                (hrmInit {})
                 (lam st595.
                    lam x942.
                      match eqi x942 2 with true
@@ -3930,8 +3930,8 @@
             in
             p_bind
               st597
-              simpleStoreSubmodel
-              (simpleInit {})
+              hrmStoreSubmodel
+              (hrmInit {})
               (lam st598.
                  lam x948.
                    match x948 with true
@@ -3994,8 +3994,8 @@
                   match
                     p_bind
                       st76
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st77.
                          lam x96.
                            match eqi x96 2 with true
@@ -4036,8 +4036,8 @@
                   match
                     p_bind
                       st81
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st82.
                          lam x104.
                            match eqi (get x104 (subi x103 1)) 2 with true
@@ -4101,8 +4101,8 @@
                   match
                     p_bind
                       st86
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st87.
                          lam x112.
                            match eqi x112 0 with true
@@ -4143,8 +4143,8 @@
                   match
                     p_bind
                       st91
-                      simpleStoreSubmodel
-                      (simpleInit {})
+                      hrmStoreSubmodel
+                      (hrmInit {})
                       (lam st92.
                          lam x120.
                            match eqi (get x120 (subi x119 1)) 0 with true
@@ -4208,8 +4208,8 @@
                 match
                   p_bind
                     st581
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st582.
                        lam x920.
                          match eqi x920 2 with true
@@ -4253,8 +4253,8 @@
                 match
                   p_bind
                     st586
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st587.
                        lam x930.
                          match eqi x930 0 with true
@@ -4295,8 +4295,8 @@
                 match
                   p_bind
                     st571
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st572.
                        lam x902.
                          match eqi (get x902 (subi x900 1)) 2 with true
@@ -4340,8 +4340,8 @@
                 match
                   p_bind
                     st576
-                    simpleStoreSubmodel
-                    (simpleInit {})
+                    hrmStoreSubmodel
+                    (hrmInit {})
                     (lam st577.
                        lam x913.
                          match eqi (get x913 (subi x912 1)) 0 with true
@@ -4490,7 +4490,7 @@
                             (st105, x138)
                           in
                           match
-                            p_subApply st105 simpleStoreSubmodel (simpleInit {}) x136 x138
+                            p_subApply st105 hrmStoreSubmodel (hrmInit {}) x136 x138
                           with
                             (st106, x139)
                           in
@@ -4682,7 +4682,7 @@
                             (st127, x172)
                           in
                           match
-                            p_subApply st127 simpleStoreSubmodel (simpleInit {}) x170 x172
+                            p_subApply st127 hrmStoreSubmodel (hrmInit {}) x170 x172
                           with
                             (st128, x173)
                           in
@@ -4766,8 +4766,8 @@
                       in
                       p_bind
                         st141
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st142.
                            lam x194.
                              match gti nextIndex11 x194 with true
@@ -5006,8 +5006,8 @@
                       in
                       p_bind
                         st170
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st171.
                            lam x241.
                              match gti nextIndex12 x241 with true
@@ -5253,8 +5253,8 @@
                       in
                       p_bind
                         st202
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st203.
                            lam x290.
                              match gti nextIndex13 x290 with true
@@ -5493,8 +5493,8 @@
                       in
                       p_bind
                         st231
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st232.
                            lam x348.
                              match gti nextIndex14 x348 with true
@@ -5766,7 +5766,9 @@
             with
               (st570, x898)
             in
-            p_assume st570 simpleStoreAssume x898
+            match p_traverseSeq st570 (lam st. lam a. (st, a)) x894 with (st570, dirtySeq) in
+            match p_map st570 (lam s. printLn (join ["Msg: ", strJoin ", " (map float2string s)])) dirtySeq with (st570, _) in
+            p_assume "anon15 assume" st570 (hrmStoreNode nodeLabel51 x893) x898
   in
   recursive
     let rec8 =
@@ -5824,8 +5826,8 @@
                     lam embeddedQMatrix11.
                       p_bind
                         st266
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st267.
                            lam x408.
                              match
@@ -5902,7 +5904,7 @@
                                in
                                match p_apply st275 x420 x422 with (st276, x423)
                                in
-                               match p_assume st276 simpleStoreAssume x423 with (st277, nextState11)
+                               match p_assume "assume here" st276 (hrmStoreBranchState nodeLabel host) x423 with (st277, nextState11)
                                in
                                match
                                  p_traverseSeq
@@ -5949,7 +5951,7 @@
                                in
                                match p_apply st281 x429 x431 with (st282, x432)
                                in
-                               match p_assume st282 simpleStoreAssume x432 with (st283, x433)
+                               match p_assume "assume here" st282 (hrmStoreBranchTime nodeLabel host) x432 with (st283, x433)
                                in
                                match
                                  p_map
@@ -6054,8 +6056,8 @@
                     lam embeddedQMatrix12.
                       p_bind
                         st291
-                        simpleStoreSubmodel
-                        (simpleInit {})
+                        hrmStoreSubmodel
+                        (hrmInit {})
                         (lam st292.
                            lam x448.
                              match
@@ -6135,7 +6137,7 @@
                                in
                                match p_apply st301 x462 x464 with (st302, x465)
                                in
-                               match p_assume st302 simpleStoreAssume x465 with (st303, nextState12)
+                               match p_assume "assume here" st302 (hrmStoreBranchState nodeLabel1 host1) x465 with (st303, nextState12)
                                in
                                match
                                  p_traverseSeq
@@ -6182,7 +6184,7 @@
                                in
                                match p_apply st307 x471 x473 with (st308, x474)
                                in
-                               match p_assume st308 simpleStoreAssume x474 with (st309, x475)
+                               match p_assume "assume here" st308 (hrmStoreBranchTime nodeLabel1 host1) x474 with (st309, x475)
                                in
                                match
                                  p_map
@@ -6364,7 +6366,7 @@
                     in
                     match p_apply st557 x878 x880 with (st558, x882)
                     in
-                    match p_assume st558 simpleStoreAssume x882 with (st559, x883)
+                    match p_assume "assume here" st558 (hrmStoreBranchTime nodeLabel12 x872) x882 with (st559, x883)
                     in
                     match
                       p_map
@@ -6410,7 +6412,7 @@
                     let st564 =
                       p_weight
                         st563
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel12 x872)
                         (lam x890.
                            externalLog
                              (addf
@@ -6499,7 +6501,7 @@
                     in
                     match p_apply st545 x860 x862 with (st546, x863)
                     in
-                    match p_assume st546 simpleStoreAssume x863 with (st547, x864)
+                    match p_assume "assume here" st546 (hrmStoreBranchTime nodeLabel11 x854) x863 with (st547, x864)
                     in
                     match
                       p_map
@@ -6545,7 +6547,7 @@
                     let st552 =
                       p_weight
                         st551
-                        simpleStoreWeight
+                        (hrmStoreBridgeSuppWeight nodeLabel11 x854)
                         (lam x871.
                            externalLog
                              (addf
@@ -7268,8 +7270,8 @@
                         match
                           p_bind
                             st364
-                            simpleStoreSubmodel
-                            (simpleInit {})
+                            hrmStoreSubmodel
+                            (hrmInit {})
                             (lam st365.
                                lam x577.
                                  match
@@ -7462,8 +7464,8 @@
                                      match
                                        p_bind
                                          st382
-                                         simpleStoreSubmodel
-                                         (simpleInit {})
+                                         hrmStoreSubmodel
+                                         (hrmInit {})
                                          (lam st383.
                                             lam x615.
                                               match gti 1 x615 with true
@@ -7572,7 +7574,7 @@
                                                 (st391, x625)
                                               in
                                               match
-                                                p_subApply st391 simpleStoreSubmodel (simpleInit {}) x623 x625
+                                                p_subApply st391 hrmStoreSubmodel (hrmInit {}) x623 x625
                                               with
                                                 (st392, x626)
                                               in
@@ -7744,10 +7746,22 @@
                         in
                         match p_apply st409 x658 x660 with (st410, x661)
                         in
+                        let nl = (match tree2 with MsgLeaf carried26
+                                             then
+                                               carried26.label
+                                             else match tree2 with MsgNode carried27
+                                             then
+                                               carried27.label
+                                             else
+                                               let #var"162" =
+                                                 print
+                                                   "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 129:90-129:100>:\nField \'label\' not found\n[0m    logWeight (branchSample.logModelDensity - branchSample.logSamplingDensity) * exp(Real([31mtree.label[0m[0m) * 0.);\n"
+                                               in
+                                               exit 1) in
                         let st411 =
                           p_weight
                             st410
-                            simpleStoreWeight
+                            (hrmStoreLikrWeight nl)
                             (/-temp-/lam x664.
                                x664)
                             x661
@@ -7844,10 +7858,22 @@
                       in
                       match bool2real st415 x665 with (st416, x666)
                       in
+                      let nl = (match tree2 with MsgLeaf carried36
+                           then
+                             carried36.label
+                           else match tree2 with MsgNode carried37
+                           then
+                             carried37.label
+                           else
+                             let #var"170" =
+                               print
+                                 "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 139:51-139:61>:\nField \'label\' not found\n[0m    let rep = suggestNodeRep(samplingProb, nHosts, [31mtree.label[0m[0m);\n"
+                             in
+                             exit 1) in
                       let st417 =
                         p_weight
                           st416
-                          simpleStoreWeight
+                          (hrmStoreNodeSuppWeight nl)
                           (lam x772.
                              externalLog
                                (addf
@@ -7922,8 +7948,8 @@
                       match
                         p_bind
                           st422
-                          simpleStoreSubmodel
-                          (simpleInit {})
+                          hrmStoreSubmodel
+                          (hrmInit {})
                           (lam st423.
                              lam x669.
                                match
@@ -8131,8 +8157,8 @@
                                    match
                                      p_bind
                                        st442
-                                       simpleStoreSubmodel
-                                       (simpleInit {})
+                                       hrmStoreSubmodel
+                                       (hrmInit {})
                                        (lam st443.
                                           lam x712.
                                             match gti 1 x712 with true
@@ -8241,7 +8267,7 @@
                                               (st451, x722)
                                             in
                                             match
-                                              p_subApply st451 simpleStoreSubmodel (simpleInit {}) x720 x722
+                                              p_subApply st451 hrmStoreSubmodel (hrmInit {}) x720 x722
                                             with
                                               (st452, x723)
                                             in
@@ -8432,10 +8458,22 @@
                       in
                       match p_apply st472 x759 x762 with (st473, x763)
                       in
+                      let nl = (match tree2 with MsgLeaf carried40
+                                         then
+                                           carried40.label
+                                         else match tree2 with MsgNode carried41
+                                         then
+                                           carried41.label
+                                         else
+                                           let #var"179" =
+                                             print
+                                               "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 157:113-157:123>:\nField \'label\' not found\n[0m    logWeight branchSample.logModelDensity - branchSample.logSamplingDensity - nodeLogSamplingDensity * exp(Real([31mtree.label[0m[0m) * 0.);\n"
+                                           in
+                                           exit 1) in
                       let st474 =
                         p_weight
                           st473
-                          simpleStoreWeight
+                          (hrmStoreLikrWeight nl)
                           (/-temp-/lam x771.
                              x771)
                           x763
@@ -8607,20 +8645,72 @@
                           repertoire = rep2 })
   in
   let input1 =
-    { dMean = 1.,
-      interactions = [ [ 2, 1 ],
-          [ 1, 2 ] ],
-      symbiontTree =
-        Node
-          { age = 1.,
-            label = 3,
-            left = Leaf
-                { age = 0., label = 1 },
-            right = Leaf
-                { age = 0., label = 2 } },
-      hostDistances = [ [ 0., 1. ],
-          [ 1., 0. ] ] }
-  in
+  { symbiontTree = Node
+    { age = 10.0
+    , label = 9
+    , left = Node
+      { label = 8
+      , age = 5.0
+      , left = Leaf
+        { label = 1
+        , age = 0.0
+        }
+      , right = Leaf
+        { label = 2
+        , age = 0.0
+        }
+      }
+    , right = Node
+      { label = 7
+      , age = 5.0
+      , left = Leaf
+        { label = 3
+        , age = 0.0
+        }
+      , right = Node
+        { label = 6
+        , age = 3.0
+        , left = Leaf
+          { label = 4
+          , age = 0.0
+          }
+        , right = Leaf
+          { label = 5
+          , age = 0.0
+          }
+        }
+      }
+    }
+  , interactions =
+      [ [0, 0, 0, 0, 2]
+      , [0, 0, 0, 2, 1]
+      , [2, 0, 1, 1, 1]
+      , [0, 1, 2, 0, 0]
+      , [0, 2, 2, 0, 0]
+      ]
+  , hostDistances =
+      [ [1., 1., 1., 1., 1.]
+      , [1., 1., 1., 1., 1.]
+      , [1., 1., 1., 1., 1.]
+      , [1., 1., 1., 1., 1.]
+      , [1., 1., 1., 1., 1.]
+      ]
+  , dMean = 1.
+  } in
+  --   { dMean = 1.,
+  --     interactions = [ [ 2, 1, 2 ],
+  --         [ 1, 2, 1 ] ],
+  --     symbiontTree =
+  --       Node
+  --         { age = 1.,
+  --           label = 3,
+  --           left = Leaf
+  --               { age = 0., label = 1 },
+  --           right = Leaf
+  --               { age = 0., label = 2 } },
+  --     hostDistances = [ [ 0., 1. ],
+  --         [ 1., 0. ] ] }
+  -- in
   match input1 with {symbiontTree = #var"X9"}
   in
   let symbiontTree = #var"X9" in
@@ -8629,21 +8719,27 @@
   let hostDistances = #var"X10" in
   let nHosts1 = length1 hostDistances in
   match
-    p_assume
-      st simpleStoreAssume (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
+    p_assume "assume here"
+      st hrmStoreLambda (p_pure (mkDirichlet [ 1., 1., 1., 1. ]))
   with
     (st482, lambda)
   in
   match
-    p_assume st482 simpleStoreAssume (p_pure (mkExponential 10.))
+    p_assume "assume here" st482 hrmStoreMu (p_pure (mkExponential 10.))
   with
     (st483, mu)
   in
   match
-    p_assume st483 simpleStoreAssume (p_pure (mkExponential 1.))
+    p_assume "assume here" st483 hrmStoreBeta (p_pure (mkExponential 1.))
   with
     (st484, beta)
   in
+  recursive let buildTree = lam stree.
+    match stree with Leaf { label = label, age = _ } then 
+      HRMLeaf { label = label }
+    else match stree with Node { label = label, age = _, left = left, right = right } in 
+      HRMNode { label = label, left = buildTree left, right = buildTree right } in
+  match p_export st484 (hrmStoreTree (buildTree symbiontTree) (match input1 with {interactions = ints} in ints)) (p_pure ()) with st484 in
   match
     p_map
       st484
@@ -8819,8 +8915,8 @@
   match
     p_bind
       st507
-      simpleStoreSubmodel
-      (simpleInit {})
+      hrmStoreSubmodel
+      (hrmInit {})
       (lam st508.
          lam x811.
            match
@@ -8889,10 +8985,22 @@
   in
   match bool2real st514 x818 with (st515, x819)
   in
+  let rl = (match postorderTree with MsgLeaf carried64
+                     then
+                       carried64.label
+                     else match postorderTree with MsgNode carried65
+                     then
+                       carried65.label
+                     else
+                       let #var"1104" =
+                         print
+                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 42:45-42:64>:\nField \'label\' not found\n[0m  weight bool2real(any(is2, rootRep)) + Real([31mpostorderTree.label[0m[0m) * 0.;\n"
+                       in
+                       exit 1) in
   let st516 =
     p_weight
       st515
-      simpleStoreWeight
+      (hrmStoreNodeSuppWeight rl)
       (lam x848.
          externalLog
            (addf
@@ -8920,10 +9028,22 @@
   with
     (st517, x820)
   in
+  let rl = (match postorderTree with MsgLeaf carried62
+                     then
+                       carried62.label
+                     else match postorderTree with MsgNode carried63
+                     then
+                       carried63.label
+                     else
+                       let #var"1103" =
+                         print
+                           "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 49:70-49:89>:\nField \'label\' not found\n[0m  logWeight (rootLogModelDensity - rootLogSamplingDensity) * exp(Real([31mpostorderTree.label[0m[0m) * 0.);\n"
+                       in
+                       exit 1) in
   let st518 =
     p_weight
       st517
-      simpleStoreWeight
+      (hrmStoreLikrWeight rl)
       (lam x847.
          mulf
            (subf
@@ -9147,44 +9267,37 @@
   with
     (st538, x845)
   in
+  let m = mapEmpty subi in
+  let makeDirtyArray = lam st. lam arr.
+    p_traverseSeq st (lam st. lam a. (st, a)) arr 
+  in
+  recursive let extractSamples = lam st. lam m. lam hTree.
+    match hTree with HistoryLeaf _ then (st, m)
+    else match hTree with HistoryNode n in
+    match makeDirtyArray st n.repertoire with (st, wrappedRep) in
+    let m = mapInsert n.label wrappedRep m in
+    match extractSamples st m n.left with (st, m) in
+    extractSamples st m n.right
+  in
+  match makeDirtyArray st538 rootRep with (st538, dRootRep) in
+  let m = mapInsert rl dRootRep m in
+  match extractSamples st538 m x841 with (st538, m) in
+  match extractSamples st538 m x843 with (st538, m) in
+  let dirtyKeyVal = lam st. lam kv. match kv with (k, v) in p_map st (lam v. (k, v)) v in 
+  match p_traverseSeq st538 dirtyKeyVal (mapToSeq m) with (st538, wrappedReps) in
+  match p_map st538 (
+    lam mu. lam beta. lam lambda. lam reps.
+      -- printLn (join ["Export root rep: [", int2string rootRep1, ", ", int2string rootRep2, "]"]);
+      { mu = mu
+      , beta = beta
+      , lambda = lambda
+      , reps = mapFromSeq subi reps 
+      }
+  ) mu with (st538, partres) in
+  match p_apply st538 partres beta with (st538, partres) in
+  match p_apply st538 partres lambda with (st538, partres) in
+  match p_apply st538 partres wrappedReps with (st538, res) in
   p_export
     st538
-    simpleStoreExport
-    (p_pure
-       (dprint
-          (ReturnType1
-             { mu = mu,
-               beta = beta,
-               tree =
-                 HistoryNode
-                   { age =
-                       match symbiontTree with Leaf carried56
-                       then
-                         carried56.age
-                       else match symbiontTree with Node carried57
-                       then
-                         carried57.age
-                       else
-                         let #var"1100" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 76:10-76:26>:\nField \'age\' not found\n[0m    age = [31msymbiontTree.age[0m[0m, label = symbiontTree.label,\n"
-                         in
-                         exit 1,
-                     label =
-                       match symbiontTree with Leaf carried58
-                       then
-                         carried58.label
-                       else match symbiontTree with Node carried59
-                       then
-                         carried59.label
-                       else
-                         let #var"1101" =
-                           print
-                             "ERROR </home/ed/treeppl-compiler-chain/treeppl/models/host-repertoire-evolution/simple-HRM.tppl 76:36-76:54>:\nField \'label\' not found\n[0m    age = symbiontTree.age, label = [31msymbiontTree.label[0m[0m,\n"
-                         in
-                         exit 1,
-                     left = x841,
-                     right = x843,
-                     history = x845,
-                     repertoire = rootRep },
-               lambda = lambda })))
+    hrmStoreExport
+    res
\ No newline at end of file
